// tempo-sdk/dist/canvas/editor.js
import {
  parseCanvasFile
} from "../chunk-6Q4I3RFN.js";

// canvas/editor.ts
import { parseSync } from "oxc-parser";
function applyCanvasSourceEdits(source, edits) {
  const sorted = [...edits].sort((left, right) => right.start - left.start);
  let previousStart = source.length + 1;
  let nextSource = source;
  for (const edit of sorted) {
    if (edit.start < 0 || edit.end < edit.start || edit.end > source.length || edit.end > previousStart) {
      throw new Error(
        `Invalid overlapping canvas source edit ${edit.start}:${edit.end}`
      );
    }
    nextSource = nextSource.slice(0, edit.start) + edit.value + nextSource.slice(edit.end);
    previousStart = edit.start;
  }
  return nextSource;
}
function updateCanvasName(options) {
  return editCanvasStringAttribute(
    options,
    "name",
    requireNonEmpty(options.name, "Canvas name"),
    (parsed) => ({
      valueSpan: parsed.nameSpan,
      attributeSpan: parsed.nameAttributeSpan
    })
  );
}
function updateCanvasPath(options) {
  return editCanvasStringAttribute(
    options,
    "path",
    requireNonEmpty(options.path, "Canvas path"),
    (parsed) => ({
      valueSpan: parsed.pathSpan,
      attributeSpan: parsed.pathAttributeSpan
    })
  );
}
function updateCanvasBackgroundColor(options) {
  return editCanvasStringAttribute(
    options,
    "backgroundColor",
    options.backgroundColor && options.backgroundColor.trim() !== "" ? options.backgroundColor : void 0,
    (parsed) => ({
      valueSpan: parsed.backgroundColorSpan,
      attributeSpan: parsed.backgroundColorAttributeSpan
    })
  );
}
function editCanvasStringAttribute(options, attributeName, value, selectSpans) {
  const parsed = parseCanvasFile(options.filePath, options.source);
  const spans = selectSpans(parsed);
  const edit = upsertJsxStringAttribute({
    source: options.source,
    openingElementSpan: requiredSpan(
      parsed.canvasOpeningElementSpan,
      "Canvas opening element not found"
    ),
    valueSpan: spans.valueSpan,
    attributeSpan: spans.attributeSpan,
    name: attributeName,
    value
  });
  return applyResult(options.source, parsed, edit ? [edit] : []);
}
function updateStoryboardName(options) {
  const parsed = parseCanvasFile(options.filePath, options.source);
  const storyboard = findStoryboard(parsed, options.storyboardId);
  if (storyboard.legacyV1) {
    const edit2 = upsertObjectStringProperty({
      source: options.source,
      objectClose: storyboard.legacyV1.objectClose,
      valueSpan: storyboard.nameSpan,
      propertySpan: storyboard.legacyV1.namePropertySpan,
      name: "name",
      value: options.name && options.name.trim() !== "" ? options.name : void 0
    });
    return applyResult(options.source, parsed, edit2 ? [edit2] : []);
  }
  const edit = upsertJsxStringAttribute({
    source: options.source,
    openingElementSpan: requireStoryboardOpeningSpan(
      storyboard,
      options.storyboardId
    ),
    valueSpan: storyboard.nameSpan,
    attributeSpan: storyboard.nameAttributeSpan,
    name: "name",
    value: options.name && options.name.trim() !== "" ? options.name : void 0
  });
  return applyResult(options.source, parsed, edit ? [edit] : []);
}
function updateStoryboardId(options) {
  const nextId = requireNonEmpty(options.nextId, "Storyboard id");
  const parsed = parseCanvasFile(options.filePath, options.source);
  const storyboard = findStoryboard(parsed, options.storyboardId);
  const duplicate = parsed.storyboards.find(
    (item) => item.storyboardKey === nextId && item.storyboardKey !== storyboard.storyboardKey
  );
  if (duplicate) {
    throw new Error(`Storyboard id "${nextId}" already exists`);
  }
  const edit = upsertJsxStringAttribute({
    source: options.source,
    openingElementSpan: requireStoryboardOpeningSpan(
      storyboard,
      options.storyboardId
    ),
    valueSpan: storyboard.idSpan,
    attributeSpan: storyboard.idAttributeSpan,
    name: "id",
    value: nextId
  });
  return applyResult(options.source, parsed, [edit]);
}
function updateStoryboardLayout(options) {
  const parsed = parseCanvasFile(options.filePath, options.source);
  const storyboard = findStoryboard(parsed, options.storyboardId);
  const spans = storyboard.layoutSpans;
  if (!spans) {
    throw new Error(
      `Storyboard "${options.storyboardId}" has no editable layout`
    );
  }
  const edits = [];
  for (const key of ["x", "y", "width", "height"]) {
    if (options.layout[key] === void 0) continue;
    const span = spans[key];
    if (!span) {
      throw new Error(
        `Storyboard "${options.storyboardId}" layout.${key} is not editable`
      );
    }
    edits.push({
      start: span.start,
      end: span.end,
      value: String(roundLayoutNumber(options.layout[key]))
    });
  }
  if (options.layout.zIndex !== void 0) {
    const value = String(roundLayoutNumber(options.layout.zIndex));
    edits.push(
      spans.zIndex ? { start: spans.zIndex.start, end: spans.zIndex.end, value } : insertLayoutProperty(options.source, spans, `zIndex: ${value}`)
    );
  }
  if (options.layout.intrinsicSizing !== void 0) {
    const value = JSON.stringify(options.layout.intrinsicSizing);
    edits.push(
      spans.intrinsicSizing ? {
        start: spans.intrinsicSizing.start,
        end: spans.intrinsicSizing.end,
        value
      } : insertLayoutProperty(
        options.source,
        spans,
        `intrinsicSizing: ${value}`
      )
    );
  }
  return applyResult(options.source, parsed, edits);
}
function updateStoryboardComponent(options) {
  const parsed = parseCanvasFile(options.filePath, options.source);
  const storyboard = findStoryboard(parsed, options.storyboardId);
  if (storyboard.kind !== "component") {
    throw new Error(
      `Storyboard "${options.storyboardId}" is not a component storyboard`
    );
  }
  const imports = [];
  if (options.componentImport) {
    const importLine = buildMissingComponentImportLine(
      parsed,
      options.source,
      options.componentImport
    );
    if (importLine) imports.push(importLine);
  }
  const importEdit = buildImportInsertionEdit(options.source, parsed, imports);
  const componentEdit = upsertJsxExpressionAttribute({
    source: options.source,
    openingElementSpan: requireStoryboardOpeningSpan(
      storyboard,
      options.storyboardId
    ),
    expressionSpan: storyboard.component?.span,
    attributeSpan: storyboard.component?.attributeSpan,
    name: "component",
    value: requireNonEmpty(
      options.componentExpression,
      "Storyboard component expression"
    )
  });
  return applyResult(options.source, parsed, [
    ...importEdit ? [importEdit] : [],
    componentEdit
  ]);
}
function updateStoryboardProps(options) {
  const parsed = parseCanvasFile(options.filePath, options.source);
  const storyboard = findStoryboard(parsed, options.storyboardId);
  const edit = upsertJsxExpressionAttribute({
    source: options.source,
    openingElementSpan: requireStoryboardOpeningSpan(
      storyboard,
      options.storyboardId
    ),
    expressionSpan: storyboard.props?.span,
    attributeSpan: storyboard.props?.attributeSpan,
    name: "props",
    value: options.propsExpression && options.propsExpression.trim() !== "" ? options.propsExpression.trim() : void 0
  });
  return applyResult(options.source, parsed, edit ? [edit] : []);
}
function updateRouteStoryboardRoute(options) {
  const parsed = parseCanvasFile(options.filePath, options.source);
  const storyboard = findStoryboard(parsed, options.storyboardId);
  if (storyboard.kind !== "route") {
    throw new Error(
      `Storyboard "${options.storyboardId}" is not a route storyboard`
    );
  }
  const edit = upsertJsxStringAttribute({
    source: options.source,
    openingElementSpan: requireStoryboardOpeningSpan(
      storyboard,
      options.storyboardId
    ),
    valueSpan: storyboard.routeSpan,
    attributeSpan: storyboard.routeAttributeSpan,
    name: "route",
    value: requireNonEmpty(options.route, "Route storyboard route")
  });
  return applyResult(options.source, parsed, [edit]);
}
function updateHtmlStoryboardSource(options) {
  const parsed = parseCanvasFile(options.filePath, options.source);
  const storyboard = findStoryboard(parsed, options.storyboardId);
  if (storyboard.kind !== "html") {
    throw new Error(
      `Storyboard "${options.storyboardId}" is not an html storyboard`
    );
  }
  const filepath = options.filepath && options.filepath.trim() !== "" ? options.filepath.trim() : void 0;
  const src = options.src && options.src.trim() !== "" ? options.src : void 0;
  if (!filepath && !src) {
    throw new Error("HTML storyboard source requires filepath or src");
  }
  const openingElementSpan = requireStoryboardOpeningSpan(
    storyboard,
    options.storyboardId
  );
  const filepathEdit = upsertJsxStringAttribute({
    source: options.source,
    openingElementSpan,
    valueSpan: storyboard.filepathSpan,
    attributeSpan: storyboard.filepathAttributeSpan,
    name: "filepath",
    value: filepath
  });
  const srcEdit = upsertJsxStringAttribute({
    source: options.source,
    openingElementSpan,
    valueSpan: storyboard.srcSpan,
    attributeSpan: storyboard.srcAttributeSpan,
    name: "src",
    value: src
  });
  return applyResult(options.source, parsed, [filepathEdit, srcEdit]);
}
function addCanvasStoryboard(options) {
  const parsed = parseCanvasFile(options.filePath, options.source);
  if (parsed.storyboards.some(
    (item) => item.storyboardKey === options.storyboard.id
  )) {
    throw new Error(`Storyboard id "${options.storyboard.id}" already exists`);
  }
  const imports = [];
  const sdkComponent = sdkComponentForStoryboard(options.storyboard.kind);
  const sdkTag = resolveSdkTag(parsed, sdkComponent);
  let storyboardTag = sdkTag.tag;
  if (sdkTag.importLine) {
    imports.push(sdkTag.importLine);
    storyboardTag = sdkTag.tag;
  }
  let componentExpression;
  if (options.storyboard.kind === "component") {
    componentExpression = options.storyboard.componentExpression ?? options.storyboard.componentImport?.localName;
    if (!componentExpression) {
      throw new Error(
        "Component storyboard requires componentExpression or componentImport"
      );
    }
    if (options.storyboard.componentImport) {
      const importLine = buildMissingComponentImportLine(
        parsed,
        options.source,
        options.storyboard.componentImport
      );
      if (importLine) imports.push(importLine);
    }
  }
  const importEdit = buildImportInsertionEdit(options.source, parsed, imports);
  const insertionEdit = buildCanvasChildInsertionEdit({
    source: options.source,
    parsed,
    tag: storyboardTag,
    storyboard: options.storyboard,
    componentExpression
  });
  const edits = [...importEdit ? [importEdit] : [], insertionEdit];
  return applyResult(options.source, parsed, edits);
}
function removeCanvasStoryboard(options) {
  const parsed = parseCanvasFile(options.filePath, options.source);
  const storyboard = findStoryboard(parsed, options.storyboardId);
  const edits = [
    {
      ...expandToFullLines(options.source, storyboard.jsxElementSpan),
      value: ""
    }
  ];
  if (options.removeUnusedComponentImport ?? true) {
    const importEdit = buildUnusedComponentImportRemoval({
      source: options.source,
      parsed,
      deletedStoryboard: storyboard,
      deletionEdit: edits[0]
    });
    if (importEdit) edits.push(importEdit);
  }
  const result = applyResult(options.source, parsed, edits);
  return {
    ...result,
    source: collapseBlankLines(result.source, detectNewline(options.source))
  };
}
function applyResult(source, parsed, edits) {
  const compactEdits = edits.filter((edit) => {
    if (!edit) return false;
    return source.slice(edit.start, edit.end) !== edit.value;
  });
  return {
    source: applyCanvasSourceEdits(source, compactEdits),
    edits: compactEdits,
    parsed
  };
}
function findStoryboard(parsed, storyboardId) {
  const storyboard = parsed.storyboards.find(
    (item) => item.storyboardKey === storyboardId
  );
  if (!storyboard) {
    throw new Error(`Storyboard "${storyboardId}" not found`);
  }
  return storyboard;
}
function requireStoryboardOpeningSpan(storyboard, storyboardId) {
  return requiredSpan(
    storyboard.openingElementSpan,
    `Storyboard "${storyboardId}" opening element not found`
  );
}
function upsertJsxStringAttribute(options) {
  if (options.value === void 0) {
    if (!options.attributeSpan) return null;
    return {
      ...expandJsxAttributeRemoval(options.source, options.attributeSpan),
      value: ""
    };
  }
  const value = JSON.stringify(options.value);
  if (options.valueSpan) {
    return {
      start: options.valueSpan.start,
      end: options.valueSpan.end,
      value
    };
  }
  const offset = findOpeningElementAttributeInsertionOffset(
    options.source,
    options.openingElementSpan
  );
  return {
    start: offset,
    end: offset,
    value: ` ${options.name}={${value}}`
  };
}
function upsertObjectStringProperty(options) {
  if (options.value === void 0) {
    if (!options.propertySpan) return null;
    return {
      ...expandObjectPropertyRemoval(options.source, options.propertySpan),
      value: ""
    };
  }
  const value = JSON.stringify(options.value);
  if (options.valueSpan) {
    return {
      start: options.valueSpan.start,
      end: options.valueSpan.end,
      value
    };
  }
  if (typeof options.objectClose !== "number") {
    throw new Error(`Object close not found for "${options.name}" property`);
  }
  return insertObjectProperty(
    options.source,
    options.objectClose,
    `${options.name}: ${value}`
  );
}
function upsertJsxExpressionAttribute(options) {
  if (options.value === void 0) {
    if (!options.attributeSpan) return null;
    return {
      ...expandJsxAttributeRemoval(options.source, options.attributeSpan),
      value: ""
    };
  }
  if (options.expressionSpan) {
    return {
      start: options.expressionSpan.start,
      end: options.expressionSpan.end,
      value: options.value
    };
  }
  const offset = findOpeningElementAttributeInsertionOffset(
    options.source,
    options.openingElementSpan
  );
  return {
    start: offset,
    end: offset,
    value: ` ${options.name}={${options.value}}`
  };
}
function insertLayoutProperty(source, spans, propertySource) {
  if (typeof spans.layoutClose !== "number") {
    throw new Error("Layout object close span not found");
  }
  return insertObjectProperty(source, spans.layoutClose, propertySource);
}
function insertObjectProperty(source, objectClose, propertySource) {
  const offset = trimWhitespaceBeforeOffset(source, objectClose);
  const previous = previousMeaningfulCharacter(source, offset);
  return {
    start: offset,
    end: offset,
    value: `${previous === "{" || previous === "," ? " " : ", "}${propertySource}`
  };
}
function formatStoryboardElement(options) {
  const newline = detectNewline(options.source);
  const closingIndent = lineIndentAt(options.source, options.insertionOffset);
  const indent = `${closingIndent}  `;
  const firstLineIndent = "  ";
  const propIndent = `${indent}  `;
  const props = storyboardElementProps(
    options.storyboard,
    options.componentExpression
  );
  return `${firstLineIndent}<${options.tag}${newline}${props.map((prop) => `${propIndent}${prop}`).join(newline)}${newline}${indent}/>${newline}${closingIndent}`;
}
function buildCanvasChildInsertionEdit(options) {
  const insertionOffset = options.parsed.canvasChildrenInsertionOffset;
  if (typeof insertionOffset === "number") {
    return {
      start: insertionOffset,
      end: insertionOffset,
      value: formatStoryboardElement({
        source: options.source,
        insertionOffset,
        tag: options.tag,
        storyboard: options.storyboard,
        componentExpression: options.componentExpression
      })
    };
  }
  const openingSpan = options.parsed.canvasOpeningElementSpan;
  if (!openingSpan) {
    throw new Error("Canvas closing element not found");
  }
  const openingSource = options.source.slice(
    openingSpan.start,
    openingSpan.end
  );
  const selfClosingMatch = openingSource.match(/\s*\/\s*>$/);
  const canvasName = openingSource.match(
    /^<\s*([A-Za-z_$][\w$]*(?:\.[A-Za-z_$][\w$]*)*)/
  )?.[1];
  if (!selfClosingMatch || !canvasName) {
    throw new Error("Canvas closing element not found");
  }
  const newline = detectNewline(options.source);
  const replacementStart = openingSpan.end - selfClosingMatch[0].length;
  const closingIndent = lineIndentAt(options.source, openingSpan.start);
  const childIndent = `${closingIndent}  `;
  const propIndent = `${childIndent}  `;
  const props = storyboardElementProps(
    options.storyboard,
    options.componentExpression
  );
  return {
    start: replacementStart,
    end: openingSpan.end,
    value: `>${newline}${childIndent}<${options.tag}${newline}${props.map((prop) => `${propIndent}${prop}`).join(
      newline
    )}${newline}${childIndent}/>${newline}${closingIndent}</${canvasName}>`
  };
}
function storyboardElementProps(storyboard, componentExpression) {
  const idProp = formatJsxStringAttribute(
    "id",
    requireNonEmpty(storyboard.id, "Storyboard id")
  );
  if (storyboard.kind === "legacy-v1") {
    return [idProp, `storyboard={${storyboard.storyboardExpression}}`];
  }
  return [
    idProp,
    ...storyboard.name && storyboard.name.trim() !== "" ? [formatJsxStringAttribute("name", storyboard.name)] : [],
    ...storyboardKindProps(storyboard, componentExpression),
    `layout={{ ${formatLayout(storyboard.layout)} }}`
  ];
}
function formatJsxStringAttribute(name, value) {
  const encoded = JSON.stringify(value);
  const plainSafe = encoded === `"${value}"` && !value.includes("&");
  return plainSafe ? `${name}=${encoded}` : `${name}={${encoded}}`;
}
function storyboardKindProps(storyboard, componentExpression) {
  if (storyboard.kind === "component") {
    return [
      `component={${componentExpression}}`,
      ...storyboard.props && storyboard.props.trim() !== "" ? [`props={${storyboard.props.trim()}}`] : []
    ];
  }
  if (storyboard.kind === "route") {
    return [formatJsxStringAttribute("route", storyboard.route)];
  }
  return [
    ...storyboard.filepath ? [formatJsxStringAttribute("filepath", storyboard.filepath)] : [],
    ...storyboard.src ? [formatJsxStringAttribute("src", storyboard.src)] : []
  ];
}
function formatLayout(layout) {
  return [
    `x: ${roundLayoutNumber(layout.x)}`,
    `y: ${roundLayoutNumber(layout.y)}`,
    `width: ${roundLayoutNumber(layout.width)}`,
    `height: ${roundLayoutNumber(layout.height)}`,
    ...layout.zIndex === void 0 ? [] : [`zIndex: ${roundLayoutNumber(layout.zIndex)}`],
    ...layout.intrinsicSizing ? [`intrinsicSizing: ${JSON.stringify(layout.intrinsicSizing)}`] : []
  ].join(", ");
}
function sdkComponentForStoryboard(kind) {
  if (kind === "route") return "RouteStoryboard";
  if (kind === "html") return "HtmlStoryboard";
  if (kind === "legacy-v1") return "LegacyStoryboardV1";
  return "Storyboard";
}
function resolveSdkTag(parsed, component) {
  for (const item of parsed.imports) {
    if (item.source !== "tempo-sdk/canvas") continue;
    if (item.kind === "namespace")
      return { tag: `${item.localName}.${component}` };
    if (item.importedName === component) return { tag: item.localName };
  }
  const localName = pickAvailableIdentifier(parsed, `TempoCanvas${component}`);
  return {
    tag: localName,
    importLine: `import { ${component} as ${localName} } from "tempo-sdk/canvas";`
  };
}
function buildMissingComponentImportLine(parsed, source, componentImport) {
  const existing = parsed.imports.find(
    (item) => item.localName === componentImport.localName
  );
  if (existing) {
    if (existing.source !== componentImport.source) {
      throw new Error(
        `Import binding "${componentImport.localName}" already exists`
      );
    }
    return null;
  }
  if (sourceContainsIdentifierNode(
    parsed.filePath,
    source,
    componentImport.localName
  )) {
    throw new Error(`Identifier "${componentImport.localName}" already exists`);
  }
  if (!componentImport.importedName || componentImport.importedName === "default") {
    return `import ${componentImport.localName} from ${JSON.stringify(
      componentImport.source
    )};`;
  }
  const specifier = componentImport.importedName === componentImport.localName ? componentImport.localName : `${componentImport.importedName} as ${componentImport.localName}`;
  return `import { ${specifier} } from ${JSON.stringify(componentImport.source)};`;
}
function buildImportInsertionEdit(source, parsed, importLines) {
  if (importLines.length === 0) return null;
  const newline = detectNewline(source);
  const uniqueLines = Array.from(new Set(importLines));
  const lastImport = parsed.imports.reduce(
    (latest, item) => !latest || item.declarationSpan.end > latest.declarationSpan.end ? item : latest,
    null
  );
  if (lastImport) {
    return {
      start: lastImport.declarationSpan.end,
      end: lastImport.declarationSpan.end,
      value: `${newline}${uniqueLines.join(newline)}`
    };
  }
  const offset = topLevelInsertionOffset(source);
  return {
    start: offset,
    end: offset,
    value: `${uniqueLines.join(newline)}${newline}`
  };
}
function buildUnusedComponentImportRemoval(options) {
  const componentImport = options.deletedStoryboard.component?.import;
  if (!componentImport || componentImport.importKind === "type") return null;
  const usedByRemainingStoryboard = options.parsed.storyboards.some(
    (storyboard) => storyboard.storyboardKey !== options.deletedStoryboard.storyboardKey && storyboard.component?.import?.localName === componentImport.localName
  );
  if (usedByRemainingStoryboard) return null;
  const sourceAfterStoryboardDelete = applyCanvasSourceEdits(options.source, [
    options.deletionEdit
  ]);
  if (componentImport.specifierCount === 1) {
    const sourceWithoutImport2 = removeSpanText(
      sourceAfterStoryboardDelete,
      componentImport.declarationSpan
    );
    if (hasIdentifier(sourceWithoutImport2, componentImport.localName))
      return null;
    return {
      ...expandToFullLines(options.source, componentImport.declarationSpan),
      value: ""
    };
  }
  if (componentImport.kind !== "named" || componentImport.hasDefaultSpecifier || componentImport.hasNamespaceSpecifier || componentImport.namedSpecifierCount <= 1) {
    return null;
  }
  const specifierRange = expandImportSpecifierRemovalRange(
    options.source,
    componentImport.span
  );
  const sourceWithoutImport = removeSpanText(
    sourceAfterStoryboardDelete,
    specifierRange
  );
  if (hasIdentifier(sourceWithoutImport, componentImport.localName))
    return null;
  return {
    ...specifierRange,
    value: ""
  };
}
function expandImportSpecifierRemovalRange(source, span) {
  const lineStart = Math.max(source.lastIndexOf("\n", span.start - 1) + 1, 0);
  const nextNewline = source.indexOf("\n", span.end);
  const lineEnd = nextNewline >= 0 ? nextNewline + 1 : source.length;
  const before = source.slice(lineStart, span.start);
  const after = source.slice(span.end, lineEnd);
  if (/^[ \t]*$/.test(before) && /^[ \t]*(,)?[ \t]*(\r?\n)?$/.test(after)) {
    return { start: lineStart, end: lineEnd };
  }
  let start = span.start;
  let end = span.end;
  while (end < source.length && /\s/.test(source[end] ?? "")) end++;
  if (source[end] === ",") {
    end++;
    while (end < source.length && (source[end] === " " || source[end] === "	")) {
      end++;
    }
    return { start, end };
  }
  while (start > 0 && (source[start - 1] === " " || source[start - 1] === "	")) {
    start--;
  }
  if (source[start - 1] === ",") {
    start--;
    while (start > 0 && (source[start - 1] === " " || source[start - 1] === "	")) {
      start--;
    }
  }
  return { start, end };
}
function removeSpanText(source, span) {
  return source.slice(0, span.start) + source.slice(span.end);
}
function findOpeningElementAttributeInsertionOffset(source, span) {
  const close = source.lastIndexOf(">", span.end - 1);
  if (close < span.start) {
    throw new Error("JSX opening element close not found");
  }
  return source[close - 1] === "/" ? close - 1 : close;
}
function expandJsxAttributeRemoval(source, span) {
  let start = span.start;
  let end = span.end;
  while (start > 0 && (source[start - 1] === " " || source[start - 1] === "	")) {
    start--;
  }
  while (end < source.length && (source[end] === " " || source[end] === "	")) {
    end++;
  }
  return { start, end };
}
function expandObjectPropertyRemoval(source, span) {
  let start = span.start;
  let end = span.end;
  while (start > 0 && /\s/.test(source[start - 1] ?? "")) start--;
  while (end < source.length && /\s/.test(source[end] ?? "")) end++;
  if (source[end] === ",") {
    end++;
    while (end < source.length && (source[end] === " " || source[end] === "	")) {
      end++;
    }
    return { start, end };
  }
  let comma = start - 1;
  while (comma >= 0 && /\s/.test(source[comma] ?? "")) comma--;
  if (source[comma] === ",") {
    start = comma;
  }
  return { start, end };
}
function expandToFullLines(source, span) {
  let start = span.start;
  let end = span.end;
  while (start > 0 && source[start - 1] !== "\n" && source[start - 1] !== "\r") {
    start--;
  }
  if (start > 0) {
    if (source[start - 1] === "\n") start--;
    if (start > 0 && source[start - 1] === "\r") start--;
  }
  while (end < source.length && (source[end] === " " || source[end] === "	")) {
    end++;
  }
  if (source.slice(end, end + 2) === "\r\n") {
    end += 2;
  } else if (source[end] === "\n" || source[end] === "\r") {
    end++;
  }
  return { start, end };
}
function expandStatementForward(source, span) {
  let end = span.end;
  while (end < source.length && (source[end] === " " || source[end] === "	")) {
    end++;
  }
  if (source.slice(end, end + 2) === "\r\n") {
    end += 2;
  } else if (source[end] === "\n" || source[end] === "\r") {
    end++;
  }
  return { start: span.start, end };
}
function lineIndentAt(source, offset) {
  const lineStart = Math.max(source.lastIndexOf("\n", offset - 1) + 1, 0);
  const linePrefix = source.slice(lineStart, offset);
  const match = linePrefix.match(/^[ \t]*/);
  return match?.[0] ?? "";
}
function previousMeaningfulCharacter(source, offset) {
  for (let index = offset - 1; index >= 0; index -= 1) {
    const character = source[index] ?? "";
    if (!/\s/.test(character)) return character;
  }
  return "";
}
function trimWhitespaceBeforeOffset(source, offset) {
  let nextOffset = offset;
  while (nextOffset > 0 && /\s/.test(source[nextOffset - 1] ?? "")) {
    nextOffset--;
  }
  return nextOffset;
}
function collapseBlankLines(source, newline) {
  return newline === "\r\n" ? source.replace(/(\r\n){3,}/g, "\r\n\r\n") : source.replace(/\n{3,}/g, "\n\n");
}
function detectNewline(source) {
  return source.includes("\r\n") ? "\r\n" : "\n";
}
function topLevelInsertionOffset(source) {
  let offset = 0;
  if (source.startsWith("#!")) {
    const newline = source.indexOf("\n");
    offset = newline >= 0 ? newline + 1 : source.length;
  }
  while (true) {
    const nextLineEnd = source.indexOf("\n", offset);
    const end = nextLineEnd >= 0 ? nextLineEnd + 1 : source.length;
    const line = source.slice(offset, end).trim();
    if (!/^["']use [^"']+["'];?$/.test(line)) break;
    offset = end;
  }
  return offset;
}
function pickAvailableIdentifier(parsed, base) {
  const usedImports = new Set(parsed.imports.map((item) => item.localName));
  if (!usedImports.has(base)) return base;
  for (let index = 2; index < 1e4; index += 1) {
    const candidate = `${base}${index}`;
    if (!usedImports.has(candidate)) return candidate;
  }
  throw new Error(`Unable to find available identifier for ${base}`);
}
function hasIdentifier(source, identifier) {
  return new RegExp(`\\b${escapeRegExp(identifier)}\\b`).test(source);
}
function sourceContainsIdentifierNode(filePath, source, name) {
  const program = parseSync(filePath, source).program;
  let found = false;
  walkWithParent(program, (node) => {
    if (found) return;
    if ((node.type === "Identifier" || node.type === "JSXIdentifier") && node.name === name) {
      found = true;
    }
  });
  return found;
}
function walkWithParent(node, visit, parent = null, key = null, seen = /* @__PURE__ */ new Set()) {
  if (!node || typeof node !== "object" || seen.has(node)) return;
  seen.add(node);
  if (typeof node.type === "string") visit(node, parent, key);
  for (const [childKey, value] of Object.entries(node)) {
    if (!value || typeof value !== "object") continue;
    if (Array.isArray(value)) {
      for (const item of value)
        walkWithParent(item, visit, node, childKey, seen);
    } else {
      walkWithParent(value, visit, node, childKey, seen);
    }
  }
}
function escapeRegExp(value) {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
function roundLayoutNumber(value) {
  if (typeof value !== "number" || !Number.isFinite(value)) {
    throw new Error("Layout values must be finite numbers");
  }
  return Math.round(value);
}
function requireNonEmpty(value, label) {
  const trimmed = value.trim();
  if (!trimmed) {
    throw new Error(`${label} must be non-empty`);
  }
  return trimmed;
}
function requiredSpan(value, message) {
  if (value === void 0) {
    throw new Error(message);
  }
  return value;
}
export {
  addCanvasStoryboard,
  applyCanvasSourceEdits,
  collapseBlankLines,
  detectNewline,
  expandStatementForward,
  expandToFullLines,
  formatJsxStringAttribute,
  hasIdentifier,
  removeCanvasStoryboard,
  sourceContainsIdentifierNode,
  topLevelInsertionOffset,
  updateCanvasBackgroundColor,
  updateCanvasName,
  updateCanvasPath,
  updateHtmlStoryboardSource,
  updateRouteStoryboardRoute,
  updateStoryboardComponent,
  updateStoryboardId,
  updateStoryboardLayout,
  updateStoryboardName,
  updateStoryboardProps,
  walkWithParent
};
