// tempo-sdk/dist/canvas/index.js
// canvas/canvas.tsx
import {
  Children,
  cloneElement,
  createContext,
  isValidElement,
  useContext,
  useMemo
} from "react";

// canvas/ids.ts
function normalizeId(value, label) {
  const normalized = value?.trim();
  if (!normalized) {
    throw new Error(`${label} is required.`);
  }
  return normalized;
}

// canvas/canvas.tsx
import { jsx } from "react/jsx-runtime";
var CanvasIdentityContext = createContext(null);
function getCanvasStoryboardRenderId(canvasPath, storyboardId) {
  return `canvas:${encodeURIComponent(
    normalizeId(canvasPath, "Canvas path")
  )}#storyboard:${encodeURIComponent(
    normalizeId(storyboardId, "Storyboard id")
  )}`;
}
function useCanvasIdentity() {
  return useContext(CanvasIdentityContext);
}
function useCanvasStoryboardRenderId(storyboardId) {
  const canvas = useCanvasIdentity();
  const normalizedId = normalizeId(storyboardId, "Storyboard id");
  return canvas ? getCanvasStoryboardRenderId(canvas.path, normalizedId) : normalizedId;
}
function Canvas({
  children,
  name,
  path,
  backgroundColor,
  defaultViewport
}) {
  const contextIdentity = useCanvasIdentity();
  const canvasName = normalizeId(name, "Canvas name");
  const canvasPath = path === void 0 ? contextIdentity?.path : normalizeId(path, "Canvas path");
  if (!canvasPath) {
    throw new Error(
      '<Canvas> must be rendered inside <CanvasContext path="..."> or receive a path prop.'
    );
  }
  const identity = useMemo(() => ({ path: canvasPath }), [canvasPath]);
  const keyedChildren = useMemo(
    () => Children.map(children, (child) => {
      if (!isValidElement(child)) return child;
      if (child.key != null) return child;
      const id = child.props.id;
      if (typeof id !== "string" || id.length === 0) return child;
      return cloneElement(child, { key: `storyboard:${id}` });
    }),
    [children]
  );
  return /* @__PURE__ */ jsx(CanvasIdentityContext.Provider, { value: identity, children: /* @__PURE__ */ jsx(
    "div",
    {
      "data-tempo-canvas": canvasName,
      "data-tempo-canvas-path": canvasPath,
      "data-tempo-canvas-background": backgroundColor,
      "data-tempo-canvas-viewport": defaultViewport ? JSON.stringify(defaultViewport) : void 0,
      children: keyedChildren
    }
  ) });
}
function CanvasContext({
  children,
  path
}) {
  const canvasPath = normalizeId(path, "Canvas path");
  const identity = useMemo(() => ({ path: canvasPath }), [canvasPath]);
  return /* @__PURE__ */ jsx(CanvasIdentityContext.Provider, { value: identity, children });
}

// canvas/canvas-host.tsx
import {
  useMemo as useMemo3
} from "react";

// canvas/runtime-context.ts
import { createContext as createContext2, useCallback, useContext as useContext2 } from "react";
var CanvasHostContext = createContext2(null);
function useRequiredCanvasHost() {
  const host = useContext2(CanvasHostContext);
  if (!host) {
    throw new Error(
      "Tempo canvas components must be rendered inside <CanvasHost>."
    );
  }
  return host;
}
function useActiveState(storyboardId) {
  const renderId = useCanvasStoryboardRenderId(storyboardId);
  const host = useRequiredCanvasHost();
  return host.getStoryboardStatus(renderId) !== "unmounted";
}
function useStoryboardStatus(storyboardId) {
  const renderId = useCanvasStoryboardRenderId(storyboardId);
  const host = useRequiredCanvasHost();
  return host.getStoryboardStatus(renderId);
}
function useStoryboardRendered(storyboardId) {
  const renderId = useCanvasStoryboardRenderId(storyboardId);
  const host = useRequiredCanvasHost();
  return useCallback(() => {
    host.markStoryboardRendered(renderId);
  }, [host, renderId]);
}
function useStoryboardErrorReporter(storyboardId) {
  const renderId = useCanvasStoryboardRenderId(storyboardId);
  const host = useRequiredCanvasHost();
  const reportError = useCallback(
    (error) => {
      host.reportStoryboardError(renderId, error);
    },
    [host, renderId]
  );
  const clearError = useCallback(() => {
    host.clearStoryboardError(renderId);
  }, [host, renderId]);
  return { reportError, clearError };
}
function useStoryboardSize(storyboardId) {
  const renderId = useCanvasStoryboardRenderId(storyboardId);
  const host = useRequiredCanvasHost();
  return host.getStoryboardSize(renderId);
}
function useStoryboardPlacement(storyboardId) {
  const renderId = useCanvasStoryboardRenderId(storyboardId);
  const host = useRequiredCanvasHost();
  return host.getStoryboardPlacement(renderId);
}

// canvas/use-canvas-host-postmessage.ts
import { useCallback as useCallback2, useEffect, useMemo as useMemo2, useRef, useState } from "react";

// canvas/canvas-host-runtime.ts
function getCanvasHostRuntime() {
  const runtime = window.__tempoCanvasHostRuntime;
  if (!runtime) {
    throw new Error("window.__tempoCanvasHostRuntime is not installed.");
  }
  return runtime;
}
function getOptionalCanvasHostRuntime() {
  return typeof window === "undefined" ? void 0 : window.__tempoCanvasHostRuntime;
}

// canvas/portal-tracking.ts
var PORTAL_MANAGED_PROPERTIES = [
  { key: "visibility", cssName: "visibility", hiddenValue: "hidden" },
  { key: "pointerEvents", cssName: "pointer-events", hiddenValue: "none" },
  {
    key: "contentVisibility",
    cssName: "content-visibility",
    hiddenValue: "hidden"
  },
  { key: "transition", cssName: "transition", hiddenValue: "none" },
  { key: "animation", cssName: "animation", hiddenValue: "none" }
];
var portalElementsByRenderId = /* @__PURE__ */ new Map();
var portalOwnerRenderIdsByElement = /* @__PURE__ */ new Map();
var portalInlineStyleByElement = /* @__PURE__ */ new Map();
var renderStatusById = /* @__PURE__ */ new Map();
var projectionPortalIdAttribute = "data-projection-portal-id";
var nextSdkPortalId = 1;
function escapeAttributeValue(value) {
  return value.replace(/\\/g, "\\\\").replace(/"/g, '\\"');
}
function ensureProjectionPortalId(portalRoot) {
  const existingId = portalRoot.getAttribute(projectionPortalIdAttribute);
  if (existingId) return existingId;
  const id = `tempo-sdk-portal-${nextSdkPortalId++}`;
  portalRoot.setAttribute(projectionPortalIdAttribute, id);
  return id;
}
function buildProjectionPortalSelector(portalRoot) {
  return `[${projectionPortalIdAttribute}="${escapeAttributeValue(
    ensureProjectionPortalId(portalRoot)
  )}"]`;
}
function findFiberKey(element) {
  return Object.keys(element).find(
    (key) => key.startsWith("__reactFiber$") || key.startsWith("__reactInternalInstance$")
  ) ?? null;
}
function findContainerKey(element) {
  return Object.keys(element).find((key) => key.startsWith("__reactContainer$")) ?? null;
}
function getFiberFromElement(element) {
  const fiberKey = findFiberKey(element);
  if (fiberKey) {
    return element[fiberKey] ?? null;
  }
  const containerKey = findContainerKey(element);
  if (containerKey) {
    return element[containerKey] ?? null;
  }
  return null;
}
var PortalTracker = class {
  constructor() {
    this.watchers = /* @__PURE__ */ new Set();
    this.ownerWatchersByPortalRoot = /* @__PURE__ */ new Map();
    this.observer = null;
    this.handleMutations = (mutations) => {
      for (const mutation of mutations) {
        for (const node of mutation.addedNodes) {
          if (node instanceof Element) {
            queueMicrotask(() => this.checkElement(node));
          }
        }
        for (const node of mutation.removedNodes) {
          if (!(node instanceof Element)) continue;
          const owners = this.ownerWatchersByPortalRoot.get(node);
          if (!owners) continue;
          for (const owner of owners) {
            owner.onPortalRemoved(node);
          }
          this.ownerWatchersByPortalRoot.delete(node);
        }
      }
    };
  }
  register(watcher) {
    this.watchers.add(watcher);
    if (this.watchers.size === 1) {
      this.startObserving();
    }
    if (typeof document !== "undefined" && document.body) {
      this.checkExistingElements();
      queueMicrotask(() => this.checkExistingElements());
    }
    return () => this.unregister(watcher);
  }
  checkExistingElements() {
    if (typeof document === "undefined" || !document.body) return;
    for (const child of Array.from(document.body.children)) {
      this.checkElement(child);
    }
  }
  unregister(watcher) {
    this.watchers.delete(watcher);
    for (const [portalRoot, owners] of this.ownerWatchersByPortalRoot) {
      owners.delete(watcher);
      if (owners.size === 0) {
        this.ownerWatchersByPortalRoot.delete(portalRoot);
      }
    }
    if (this.watchers.size === 0) {
      this.stopObserving();
    }
  }
  startObserving() {
    if (this.observer || typeof document === "undefined" || !document.body || typeof MutationObserver === "undefined") {
      return;
    }
    this.observer = new MutationObserver(this.handleMutations);
    this.observer.observe(document.body, { childList: true });
  }
  stopObserving() {
    this.observer?.disconnect();
    this.observer = null;
  }
  checkElement(element) {
    const ancestorDomNodes = this.getDomAncestorsViaFiber(element);
    if (ancestorDomNodes.length === 0) return;
    let owners = this.ownerWatchersByPortalRoot.get(element);
    for (const watcher of this.watchers) {
      const belongsToWatcher = ancestorDomNodes.some(
        (node) => watcher.element.contains(node) || watcher.element === node
      );
      if (!belongsToWatcher) continue;
      if (!owners) {
        owners = /* @__PURE__ */ new Set();
        this.ownerWatchersByPortalRoot.set(element, owners);
      }
      if (owners.has(watcher)) continue;
      owners.add(watcher);
      watcher.onPortalFound(element);
    }
  }
  getDomAncestorsViaFiber(element) {
    const fiber = getFiberFromElement(element);
    if (!fiber) return [];
    const domNodes = [];
    let current = fiber.return;
    while (current) {
      if (current.stateNode instanceof Element) {
        domNodes.push(current.stateNode);
      }
      current = current.return;
    }
    return domNodes;
  }
};
var portalTracker = null;
function getPortalTracker() {
  if (!portalTracker) {
    portalTracker = new PortalTracker();
  }
  return portalTracker;
}
function getStylablePortalElement(portalRoot) {
  if (portalRoot instanceof HTMLElement || portalRoot instanceof SVGElement) {
    return portalRoot;
  }
  return null;
}
function ensurePortalInlineStyle(portalRoot) {
  const existing = portalInlineStyleByElement.get(portalRoot);
  if (existing) return existing;
  const stylableElement = getStylablePortalElement(portalRoot);
  const baseline = {
    visibility: readInlineStyleValue(stylableElement, "visibility"),
    pointerEvents: readInlineStyleValue(stylableElement, "pointer-events"),
    contentVisibility: readInlineStyleValue(
      stylableElement,
      "content-visibility"
    ),
    transition: readInlineStyleValue(stylableElement, "transition"),
    animation: readInlineStyleValue(stylableElement, "animation")
  };
  portalInlineStyleByElement.set(portalRoot, baseline);
  return baseline;
}
function readInlineStyleValue(element, propertyName) {
  return {
    value: element?.style.getPropertyValue(propertyName) ?? "",
    priority: element?.style.getPropertyPriority(propertyName) ?? ""
  };
}
function restoreInlineStyleValue(style, propertyName, baseline) {
  if (baseline.value) {
    style.setProperty(propertyName, baseline.value, baseline.priority);
    return;
  }
  style.removeProperty(propertyName);
}
function restorePortalInlineStyle(portalRoot) {
  const stylableElement = getStylablePortalElement(portalRoot);
  const baseline = portalInlineStyleByElement.get(portalRoot);
  if (!stylableElement || !baseline) return;
  for (const property of PORTAL_MANAGED_PROPERTIES) {
    restoreInlineStyleValue(
      stylableElement.style,
      property.cssName,
      baseline[property.key]
    );
  }
}
function syncPortalVisibility(portalRoot) {
  const owners = portalOwnerRenderIdsByElement.get(portalRoot);
  if (!owners || owners.size === 0) {
    restorePortalInlineStyle(portalRoot);
    portalInlineStyleByElement.delete(portalRoot);
    return;
  }
  const stylableElement = getStylablePortalElement(portalRoot);
  if (!stylableElement) return;
  const shouldBeVisible = [...owners].some(
    (renderId) => renderStatusById.get(renderId) === "active"
  );
  if (shouldBeVisible) {
    restorePortalInlineStyle(portalRoot);
    portalInlineStyleByElement.delete(portalRoot);
    return;
  }
  ensurePortalInlineStyle(portalRoot);
  for (const property of PORTAL_MANAGED_PROPERTIES) {
    stylableElement.style.setProperty(
      property.cssName,
      property.hiddenValue,
      "important"
    );
  }
}
function trackPortal(renderId, portalRoot) {
  ensureProjectionPortalId(portalRoot);
  const portals = portalElementsByRenderId.get(renderId) ?? /* @__PURE__ */ new Set();
  portals.add(portalRoot);
  portalElementsByRenderId.set(renderId, portals);
  const owners = portalOwnerRenderIdsByElement.get(portalRoot) ?? /* @__PURE__ */ new Set();
  owners.add(renderId);
  portalOwnerRenderIdsByElement.set(portalRoot, owners);
  syncPortalVisibility(portalRoot);
}
function untrackPortal(renderId, portalRoot) {
  const portals = portalElementsByRenderId.get(renderId);
  if (!portals) return;
  portals.delete(portalRoot);
  if (portals.size === 0) {
    portalElementsByRenderId.delete(renderId);
  }
  const owners = portalOwnerRenderIdsByElement.get(portalRoot);
  if (owners) {
    owners.delete(renderId);
    if (owners.size === 0) {
      portalOwnerRenderIdsByElement.delete(portalRoot);
    }
  }
  syncPortalVisibility(portalRoot);
}
function releaseStoryboardPortals(renderId) {
  const portals = portalElementsByRenderId.get(renderId);
  if (!portals) {
    renderStatusById.delete(renderId);
    return;
  }
  for (const portalRoot of [...portals]) {
    untrackPortal(renderId, portalRoot);
  }
  renderStatusById.delete(renderId);
}
function getStoryboardPortalSelectors(renderId) {
  const portals = portalElementsByRenderId.get(renderId);
  if (!portals) return [];
  return [...portals].filter((portalRoot) => portalRoot.isConnected).map(buildProjectionPortalSelector);
}
function syncStoryboardPortals(renderId, status) {
  renderStatusById.set(renderId, status);
  const portals = portalElementsByRenderId.get(renderId);
  if (!portals) return;
  for (const portalRoot of portals) {
    syncPortalVisibility(portalRoot);
  }
}
function trackStoryboardPortals(renderId, element, status) {
  syncStoryboardPortals(renderId, status);
  const unregister = getPortalTracker().register({
    element,
    onPortalFound(portalRoot) {
      trackPortal(renderId, portalRoot);
    },
    onPortalRemoved(portalRoot) {
      untrackPortal(renderId, portalRoot);
    }
  });
  return () => {
    unregister();
    releaseStoryboardPortals(renderId);
  };
}

// canvas/canvas-host-operations.ts
var escapeRenderId = (renderId) => renderId.replace(/\\/g, "\\\\").replace(/"/g, '\\"');
var buildRenderSelector = (renderId) => `[data-render-id="${escapeRenderId(renderId)}"]`;
var buildRenderContentSelector = (renderId) => `[data-render-content="${escapeRenderId(renderId)}"]`;
var buildRenderContentFiberSelector = (renderId) => `${buildRenderContentSelector(renderId)} > :not([data-tempo-hide-fiber="true"])`;
function stripRenderContentWrapperAttributes(capture, renderId) {
  if (!capture || typeof capture !== "object") return capture;
  const dom = capture.dom;
  if (!dom) return capture;
  const shouldStripAttributes = dom.selector === "" && dom.attributes?.["data-render-content"] === renderId;
  const shouldStripRootAttributes = dom.rootAttributes?.["data-render-content"] === renderId;
  if (!shouldStripAttributes && !shouldStripRootAttributes) return capture;
  return {
    ...capture,
    dom: {
      ...dom,
      attributes: shouldStripAttributes ? void 0 : dom.attributes,
      rootAttributes: shouldStripRootAttributes ? void 0 : dom.rootAttributes
    }
  };
}
function stripRenderContentWrapperFromCombinedCapture(capture, renderId) {
  if (!capture || typeof capture !== "object" || !("html" in capture)) {
    return capture;
  }
  const html = capture.html;
  if (!html) return capture;
  return {
    ...capture,
    html: stripRenderContentWrapperAttributes(html, renderId)
  };
}
function captureStoryboardImage(renderId, options) {
  const portalSelectors = [
    ...options?.portalSelectors ?? [],
    ...getStoryboardPortalSelectors(renderId)
  ];
  return getCanvasHostRuntime().captureSelector(
    buildRenderSelector(renderId),
    portalSelectors.length > 0 ? { ...options, portalSelectors: [...new Set(portalSelectors)] } : options
  );
}
async function captureStoryboardHtml(renderId, options) {
  const capture = await getCanvasHostRuntime().captureHtml?.({
    selector: buildRenderContentSelector(renderId),
    timeout: options?.timeout
  });
  return capture ? stripRenderContentWrapperAttributes(capture, renderId) : null;
}
function captureStoryboardStaticHtml(renderId, options) {
  return getCanvasHostRuntime().captureStaticHtml?.({
    selector: buildRenderContentSelector(renderId),
    timeout: options?.timeout
  }) ?? Promise.resolve(null);
}
async function captureStoryboardCombined(renderId, options) {
  const capture = await getCanvasHostRuntime().captureCombined?.({
    selector: buildRenderContentSelector(renderId),
    fiberSelector: options?.fiberSelector ?? buildRenderContentFiberSelector(renderId),
    fiberTags: options?.fiberTags,
    timeout: options?.timeout
  });
  return capture ? stripRenderContentWrapperFromCombinedCapture(capture, renderId) : null;
}
function watchStoryboardHtml(renderId, options) {
  const runtime = getCanvasHostRuntime();
  if (!runtime.watchHtml) {
    return Promise.reject(new Error("watchHtml is not available."));
  }
  return runtime.watchHtml({
    selector: buildRenderContentSelector(renderId),
    onUpdate(capture) {
      options.onUpdate(stripRenderContentWrapperAttributes(capture, renderId));
    },
    onError: options.onError,
    debounceDelay: options.debounceDelay,
    maxWait: options.maxWait,
    timeout: options.timeout
  });
}
function watchStoryboardCombined(renderId, options) {
  const runtime = getCanvasHostRuntime();
  if (!runtime.watchCombined) {
    return Promise.reject(new Error("watchCombined is not available."));
  }
  return runtime.watchCombined({
    selector: buildRenderContentSelector(renderId),
    fiberSelector: options.fiberSelector ?? buildRenderContentFiberSelector(renderId),
    fiberTags: options.fiberTags,
    onUpdate(capture) {
      options.onUpdate(
        stripRenderContentWrapperFromCombinedCapture(capture, renderId)
      );
    },
    onError: options.onError,
    debounceDelay: options.debounceDelay,
    maxWait: options.maxWait,
    timeout: options.timeout,
    knownSourceVersion: options.knownSourceVersion
  });
}

// canvas/canvas-host-protocol.ts
function isRecord(value) {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
function readCanvasHostCommand(message) {
  if (!isRecord(message) || typeof message.type !== "string") return null;
  return {
    type: message.type,
    requestId: message.requestId,
    payload: isRecord(message.payload) ? message.payload : {}
  };
}
var readTrimmedString = (value) => {
  if (typeof value !== "string") return void 0;
  const normalized = value.trim();
  return normalized.length > 0 ? normalized : void 0;
};
function resolveRenderTarget(payload) {
  const explicitRenderId = readTrimmedString(payload.renderId);
  const canvasPath = readTrimmedString(payload.canvasPath);
  const storyboardId = readTrimmedString(payload.storyboardId);
  const renderId = explicitRenderId || (canvasPath && storyboardId ? getCanvasStoryboardRenderId(canvasPath, storyboardId) : storyboardId);
  return renderId ? {
    renderId,
    ...canvasPath ? { canvasPath } : {},
    ...storyboardId ? { storyboardId } : {}
  } : null;
}
function requireRenderTarget(commandType, payload) {
  const target = resolveRenderTarget(payload);
  if (!target) {
    throw new Error(
      `${commandType} requires payload.renderId or payload.canvasPath + payload.storyboardId`
    );
  }
  return target;
}
var readPositiveFiniteNumber = (value, fieldName) => {
  if (typeof value !== "number" || !Number.isFinite(value) || value <= 0) {
    throw new Error(`${fieldName} must be a positive finite number`);
  }
  return value;
};
function readOptionalNumber(payload, fieldName) {
  return typeof payload[fieldName] === "number" ? payload[fieldName] : void 0;
}
function readOptionalSize(payload) {
  const hasWidth = payload.width != null;
  const hasHeight = payload.height != null;
  if (!hasWidth && !hasHeight) return {};
  if (!hasWidth || !hasHeight) {
    throw new Error("width and height must be provided together");
  }
  return {
    width: readPositiveFiniteNumber(payload.width, "width"),
    height: readPositiveFiniteNumber(payload.height, "height")
  };
}
function readRenderRequest(commandType, payload) {
  const target = requireRenderTarget(commandType, payload);
  const importPath = readTrimmedString(payload.importPath);
  const importName = readTrimmedString(payload.importName);
  const props = isRecord(payload.props) ? payload.props : void 0;
  return {
    ...target,
    importPath: importPath ?? target.canvasPath ?? target.renderId,
    importName: importName ?? target.storyboardId ?? target.renderId,
    ...props ? { props } : {},
    ...readOptionalSize(payload),
    intrinsicSize: payload.intrinsicSize === true ? true : void 0
  };
}
function readWatchId(commandType, payload) {
  const watchId = readTrimmedString(payload.watchId);
  if (!watchId) {
    throw new Error(`${commandType} requires payload.watchId`);
  }
  return watchId;
}

// canvas/canvas-host-render-state.ts
var propsMatch = (a, b) => JSON.stringify(a ?? {}) === JSON.stringify(b ?? {});
function renderRequestMatchesEntry(entry, request) {
  return entry?.importPath === request.importPath && entry.importName === request.importName && entry.width === request.width && entry.height === request.height && entry.intrinsicSize === request.intrinsicSize && propsMatch(entry.props, request.props);
}
function acquireRender(entries, request) {
  const next = new Map(entries);
  const current = next.get(request.renderId);
  const changed = !renderRequestMatchesEntry(current, request);
  next.set(request.renderId, {
    ...request,
    status: "active",
    activeCount: (current?.activeCount ?? 0) + 1,
    revision: changed ? (current?.revision ?? 0) + 1 : current?.revision ?? 1
  });
  return next;
}
function releaseRender(entries, renderId, maxCached, now = Date.now()) {
  const current = entries.get(renderId);
  if (!current) return new Map(entries);
  const next = new Map(entries);
  const activeCount = Math.max(0, current.activeCount - 1);
  if (activeCount > 0) {
    next.set(renderId, { ...current, activeCount, status: "active" });
    return next;
  }
  next.set(renderId, {
    ...current,
    activeCount: 0,
    status: "cached",
    lastInactiveAt: now
  });
  return evictCachedRenders(next, maxCached);
}
function evictCachedRenders(entries, maxCached) {
  const next = new Map(entries);
  const cached = [...next.values()].filter((entry) => entry.status === "cached").sort((a, b) => (a.lastInactiveAt ?? 0) - (b.lastInactiveAt ?? 0));
  for (const entry of cached.slice(0, Math.max(0, cached.length - maxCached))) {
    next.delete(entry.renderId);
  }
  return next;
}
function buildRenderStatePayload(entries, maxCached) {
  return {
    renders: [...entries.values()].map(
      ({
        activeCount: _activeCount,
        lastInactiveAt: _lastInactiveAt,
        revision: _revision,
        ...entry
      }) => entry
    ),
    maxCached
  };
}
function buildRenderStatusMap(entries) {
  return new Map([...entries].map(([id, entry]) => [id, entry.status]));
}
function buildRenderSizeMap(entries) {
  return new Map(
    [...entries].map(([id, entry]) => [
      id,
      {
        width: entry.width,
        height: entry.height,
        intrinsicSize: entry.intrinsicSize,
        revision: entry.revision
      }
    ])
  );
}

// canvas/runtime-problems.ts
var CanvasHostProblemTracker = class {
  constructor() {
    this.hostProblemHmrGeneration = 0;
    this.componentProblems = /* @__PURE__ */ new Map();
    this.componentProblemHmrGenerations = /* @__PURE__ */ new Map();
    this.canvasProblems = /* @__PURE__ */ new Map();
    this.watchers = /* @__PURE__ */ new Set();
    this.hostProblemClearedWatchers = /* @__PURE__ */ new Set();
    this.hmrGeneration = 0;
  }
  snapshot() {
    const currentViteOverlay = readCurrentViteOverlay();
    const hostProblem = currentViteOverlay ?? (this.hostProblem?.sourceKind === "vite-overlay" ? void 0 : this.hostProblem);
    return {
      hostProblem,
      componentProblems: [...this.componentProblems].map(
        ([renderId, problem]) => ({ renderId, problem })
      ),
      canvasProblems: [...this.canvasProblems].map(([canvasPath, problem]) => ({
        canvasPath,
        problem
      }))
    };
  }
  watch(callback) {
    this.watchers.add(callback);
    return () => {
      this.watchers.delete(callback);
    };
  }
  watchHostProblemCleared(callback) {
    this.hostProblemClearedWatchers.add(callback);
    return () => {
      this.hostProblemClearedWatchers.delete(callback);
    };
  }
  reportHostProblem(sourceKind, error, details = {}) {
    this.hostProblem = createProblem(sourceKind, error, details);
    this.hostProblemHmrGeneration = this.hmrGeneration;
    this.publish();
  }
  reportProblemNearRender(sourceKind, error, details = {}) {
    const problem = createProblem(sourceKind, error, details);
    const renderId = findRenderIdForProblem(problem);
    if (renderId) {
      this.componentProblems.set(renderId, problem);
      this.componentProblemHmrGenerations.set(renderId, this.hmrGeneration);
    } else {
      this.hostProblem = problem;
      this.hostProblemHmrGeneration = this.hmrGeneration;
    }
    this.publish();
  }
  // A whole-canvas failure (its module failed to import, so none of its
  // storyboards ever registered). Keyed by canvas path, not renderId — the
  // renderer scopes it to that canvas's storyboards.
  reportCanvasProblem(canvasPath, sourceKind, error, details = {}) {
    this.canvasProblems.set(canvasPath, createProblem(sourceKind, error, details));
    this.publish();
  }
  clearCanvasProblem(canvasPath) {
    if (!this.canvasProblems.delete(canvasPath)) return;
    this.publish();
  }
  clearHostProblem(sourceKinds) {
    if (!this.hostProblem) return;
    if (sourceKinds && !sourceKinds.includes(this.hostProblem.sourceKind)) return;
    const clearedProblem = this.hostProblem;
    this.hostProblem = void 0;
    this.hostProblemHmrGeneration = 0;
    this.publish();
    this.publishHostProblemCleared(clearedProblem);
  }
  reportComponentProblem(renderId, sourceKind, error, details = {}) {
    this.componentProblems.set(
      renderId,
      createProblem(sourceKind, error, details)
    );
    this.componentProblemHmrGenerations.set(renderId, this.hmrGeneration);
    this.publish();
  }
  clearComponentProblem(renderId, sourceKinds) {
    const problem = this.componentProblems.get(renderId);
    if (!problem) return;
    if (sourceKinds && !sourceKinds.includes(problem.sourceKind)) return;
    this.componentProblems.delete(renderId);
    this.componentProblemHmrGenerations.delete(renderId);
    this.publish();
  }
  hasComponentProblem(renderId) {
    return this.componentProblems.has(renderId);
  }
  clearRecoveredComponentProblemsAfterRender(renderId) {
    let changed = false;
    const problem = this.componentProblems.get(renderId);
    if (problem && problem.sourceKind !== "render-error-boundary" && this.hmrGeneration > (this.componentProblemHmrGenerations.get(renderId) ?? this.hmrGeneration)) {
      this.componentProblems.delete(renderId);
      this.componentProblemHmrGenerations.delete(renderId);
      changed = true;
    }
    if (this.clearRecoveredHostProblemAfterRender(renderId)) changed = true;
    if (changed) this.publish();
  }
  noteHmrUpdateStarted() {
    this.hmrGeneration += 1;
  }
  dispose() {
    this.watchers.clear();
    this.hostProblemClearedWatchers.clear();
  }
  inspectCurrentProblems(options = {}) {
    const renderIdFilter = options.renderId ?? void 0;
    const currentRenderFailures = [];
    for (const [renderId, failure] of this.componentProblems) {
      if (renderIdFilter != null && renderId !== renderIdFilter) continue;
      currentRenderFailures.push({ renderId, failure });
    }
    const currentHostFailure = readCurrentViteOverlay() ?? this.hostProblem;
    return {
      ...currentHostFailure ? { currentHostFailure } : {},
      currentRenderFailures,
      currentCanvasFailures: [...this.canvasProblems].map(
        ([canvasPath, failure]) => ({ canvasPath, failure })
      ),
      observedAt: Date.now()
    };
  }
  publish() {
    for (const watcher of this.watchers) {
      watcher();
    }
  }
  publishHostProblemCleared(problem) {
    for (const watcher of this.hostProblemClearedWatchers) {
      watcher(problem);
    }
  }
  clearRecoveredHostProblemAfterRender(renderId) {
    if (!this.hostProblem || !renderRecoveredHostProblemSourceKinds.includes(this.hostProblem.sourceKind)) {
      return false;
    }
    if (this.hmrGeneration <= this.hostProblemHmrGeneration) return false;
    if (this.hostProblem.sourceKind === "vite-hmr-error" && findRenderIdForProblem(this.hostProblem) !== renderId) {
      return false;
    }
    const clearedProblem = this.hostProblem;
    this.hostProblem = void 0;
    this.hostProblemHmrGeneration = 0;
    this.publishHostProblemCleared(clearedProblem);
    return true;
  }
};
function createCanvasHostProblemTracker() {
  return new CanvasHostProblemTracker();
}
function installCanvasHostRuntimeProblemTracking(tracker) {
  if (typeof window === "undefined") return () => {
  };
  const cleanupHostCapture = installHostProblemCapture(tracker);
  const cleanupCanvasImportCapture = installCanvasImportProblemCapture(tracker);
  const cleanupViteCapture = installViteProblemCapture(tracker);
  const cleanupInspector = installCurrentProblemInspector(tracker);
  return () => {
    cleanupHostCapture();
    cleanupCanvasImportCapture();
    cleanupViteCapture();
    cleanupInspector();
    tracker.dispose();
  };
}
function installCanvasImportProblemCapture(tracker) {
  const onCanvasImportError = (event) => {
    const detail = event.detail;
    if (typeof detail !== "object" || detail === null) return;
    const canvasPath = typeof detail.canvasPath === "string" ? detail.canvasPath : "";
    if (!canvasPath) return;
    const message = typeof detail.message === "string" && detail.message ? detail.message : "Canvas module failed to load";
    tracker.reportCanvasProblem(canvasPath, "runtime-bootstrap", message, {
      stack: typeof detail.stack === "string" ? detail.stack : void 0,
      filename: canvasPath
    });
  };
  window.addEventListener("tempo:canvas-import-error", onCanvasImportError);
  return () => {
    window.removeEventListener("tempo:canvas-import-error", onCanvasImportError);
  };
}
var renderRecoveredHostProblemSourceKinds = ["page-error", "unhandled-rejection", "console-vite-error", "vite-hmr-error"];
function createProblem(sourceKind, error, details = {}) {
  return {
    sourceKind,
    message: error instanceof Error ? error.message : String(error),
    stack: error instanceof Error ? error.stack : void 0,
    frameUrl: typeof window === "undefined" ? void 0 : window.location.href,
    timestamp: Date.now(),
    ...details
  };
}
function findRenderIdForProblem(problem) {
  if (typeof document === "undefined") return void 0;
  const haystack = [
    problem.message,
    problem.stack,
    problem.componentStack,
    problem.filename
  ].filter(Boolean).join("\n");
  if (!haystack) return void 0;
  for (const slot of document.querySelectorAll("[data-render-id]")) {
    const renderId = slot.getAttribute("data-render-id");
    if (!renderId) continue;
    const sourceFiles = getObservedSourceFiles(slot);
    if (sourceFiles.some((sourceFile) => problemTextMentionsSource(haystack, sourceFile))) {
      return renderId;
    }
  }
  return void 0;
}
var sourceAnnotationSelector = "[data-tempo-source-file], [data-tempo-source-loc], [data-tempo-filepath], [data-tempo-html-filepath]";
function getObservedSourceFiles(slot) {
  const sourceFiles = /* @__PURE__ */ new Set();
  for (const element of slot.querySelectorAll(sourceAnnotationSelector)) {
    for (const attribute of [
      "data-tempo-source-file",
      "data-tempo-filepath",
      "data-tempo-html-filepath"
    ]) {
      const value = element.getAttribute(attribute);
      if (value) sourceFiles.add(value);
    }
    const sourceLoc = element.getAttribute("data-tempo-source-loc");
    const sourceFile = sourceLoc?.split(":")[0];
    if (sourceFile) sourceFiles.add(sourceFile);
  }
  return [...sourceFiles];
}
function problemTextMentionsSource(text, sourceFile) {
  const normalizedSource = sourceFile.replace(/\\/g, "/");
  const basename = normalizedSource.split("/").pop() ?? normalizedSource;
  return text.includes(normalizedSource) || basename !== "" && text.includes(basename);
}
function installHostProblemCapture(tracker) {
  const onError = (event) => {
    if (event.target instanceof Element) return;
    tracker.reportProblemNearRender("page-error", event.error ?? event.message, {
      filename: event.filename || void 0,
      line: typeof event.lineno === "number" ? event.lineno : void 0,
      column: typeof event.colno === "number" ? event.colno : void 0
    });
  };
  const onUnhandledRejection = (event) => {
    tracker.reportProblemNearRender("unhandled-rejection", event.reason);
  };
  const resourceTags = /* @__PURE__ */ new Set([
    "IMG",
    "SCRIPT",
    "LINK",
    "AUDIO",
    "VIDEO",
    "SOURCE"
  ]);
  const onResourceError = (event) => {
    const target = event.target;
    if (!(target instanceof Element) || !resourceTags.has(target.tagName)) {
      return;
    }
    const url = target.getAttribute("src") || target.getAttribute("href") || "";
    const message = "Failed to load " + target.tagName.toLowerCase() + (url ? ": " + url : "");
    const renderId = target.closest("[data-render-id]")?.getAttribute("data-render-id");
    if (renderId) {
      tracker.reportComponentProblem(renderId, "resource-load", message);
    } else {
      tracker.reportHostProblem("resource-load", message, {
        filename: url || void 0
      });
    }
  };
  const onFontLoadingError = (event) => {
    const faces = event.fontfaces ?? [];
    const families = faces.map((face) => face.family).filter(Boolean).join(", ");
    tracker.reportHostProblem(
      "resource-load",
      "Failed to load font" + (families ? ": " + families : "")
    );
  };
  const fontFaceSet = typeof document.fonts?.addEventListener === "function" ? document.fonts : void 0;
  window.addEventListener("error", onError);
  window.addEventListener("unhandledrejection", onUnhandledRejection);
  window.addEventListener("error", onResourceError, { capture: true });
  document.addEventListener("error", onResourceError, { capture: true });
  fontFaceSet?.addEventListener("loadingerror", onFontLoadingError);
  return () => {
    window.removeEventListener("error", onError);
    window.removeEventListener("unhandledrejection", onUnhandledRejection);
    window.removeEventListener("error", onResourceError, { capture: true });
    document.removeEventListener("error", onResourceError, { capture: true });
    fontFaceSet?.removeEventListener("loadingerror", onFontLoadingError);
  };
}
function installViteProblemCapture(tracker) {
  const reconcileViteOverlay = () => {
    const problem = readCurrentViteOverlay();
    if (problem) {
      tracker.reportHostProblem("vite-overlay", problem.message, {
        stack: problem.stack,
        filename: problem.filename
      });
    } else {
      tracker.clearHostProblem(["vite-overlay"]);
    }
  };
  const observer = new MutationObserver(reconcileViteOverlay);
  if (document.body) {
    observer.observe(document.body, { childList: true });
    reconcileViteOverlay();
    void Promise.resolve().then(reconcileViteOverlay);
  }
  const viteHmrFailedModules = /* @__PURE__ */ new Map();
  const latestViteHmrFailure = () => {
    let latest;
    for (const problem of viteHmrFailedModules.values()) {
      if (!latest || problem.timestamp >= latest.timestamp) latest = problem;
    }
    return latest;
  };
  const reconcileViteHmrHostProblem = () => {
    const latest = latestViteHmrFailure();
    if (latest) {
      tracker.reportHostProblem("vite-hmr-error", latest.message, latest);
    } else {
      tracker.clearHostProblem(["vite-hmr-error"]);
    }
  };
  const failedReloadPattern = /^Failed to reload (.+?)\. This could be due to/;
  const pathlessViteErrorKey = "__tempo:pathless-vite-error";
  let viteHmrCycleActive = false;
  let cycleFailedPaths = /* @__PURE__ */ new Map();
  let lastCycleViteErrorDetail;
  const onViteBeforeUpdate = () => {
    viteHmrCycleActive = true;
    cycleFailedPaths = /* @__PURE__ */ new Map();
    lastCycleViteErrorDetail = void 0;
    tracker.noteHmrUpdateStarted();
  };
  const onViteAfterUpdate = (payload) => {
    const updatedPaths = readUpdatePaths(payload);
    let shouldReconcile = false;
    if (updatedPaths.length > 0 && !cycleFailedPaths.has(pathlessViteErrorKey)) {
      shouldReconcile = viteHmrFailedModules.delete(pathlessViteErrorKey);
    }
    for (const [path, problem] of cycleFailedPaths) {
      viteHmrFailedModules.set(path, problem);
      shouldReconcile = true;
    }
    viteHmrCycleActive = false;
    cycleFailedPaths = /* @__PURE__ */ new Map();
    if (shouldReconcile) reconcileViteHmrHostProblem();
  };
  const onViteError = (payload) => {
    const error = readViteError(payload);
    if (!error) return;
    lastCycleViteErrorDetail = error;
    const path = error.filename ?? pathlessViteErrorKey;
    const problem = createProblem("vite-hmr-error", error.message, {
      stack: error.stack,
      filename: error.filename
    });
    if (viteHmrCycleActive) {
      cycleFailedPaths.set(path, problem);
    } else {
      viteHmrFailedModules.set(path, problem);
    }
    reconcileViteHmrHostProblem();
  };
  let hotEventsCancelled = false;
  let subscribedHot;
  const subscribeHotEvents = (hot2) => {
    subscribedHot = hot2;
    hot2.on?.("vite:beforeUpdate", onViteBeforeUpdate);
    hot2.on?.("vite:afterUpdate", onViteAfterUpdate);
    hot2.on?.("vite:error", onViteError);
  };
  const hot = readViteHot();
  if (hot?.on) {
    subscribeHotEvents(hot);
  } else {
    void import(
      /* @vite-ignore */
      /* webpackIgnore: true */
      /* turbopackIgnore: true */
      "/@vite/client"
    ).then((client) => {
      const createHotContext = client.createHotContext;
      const context = createHotContext?.("/__tempo-canvas-host__");
      if (!context || hotEventsCancelled) return;
      subscribeHotEvents(context);
    }).catch(() => {
    });
  }
  const stopClearedWatch = tracker.watchHostProblemCleared((problem) => {
    if (problem.sourceKind !== "vite-hmr-error") return;
    const key = problem.filename ?? pathlessViteErrorKey;
    if (!viteHmrFailedModules.has(key)) return;
    viteHmrFailedModules.delete(key);
    void Promise.resolve().then(reconcileViteHmrHostProblem);
  });
  const originalConsoleError = console.error;
  const patchedConsoleError = (...args) => {
    originalConsoleError(...args);
    const error = readViteConsoleError(args[0], args[1]);
    if (!error) return;
    const failedReload = failedReloadPattern.exec(error.message);
    if (failedReload) {
      const path = failedReload[1];
      const detail = lastCycleViteErrorDetail;
      const problem = createProblem(
        "vite-hmr-error",
        detail?.message ?? "Failed to reload " + path,
        {
          stack: detail?.stack ?? error.stack,
          filename: detail?.filename ?? path
        }
      );
      if (detail) {
        cycleFailedPaths.set(path, problem);
        viteHmrFailedModules.set(path, problem);
        reconcileViteHmrHostProblem();
      } else {
        void verifyModuleStillExists(path).then((exists) => {
          if (exists) viteHmrFailedModules.set(path, problem);
          else viteHmrFailedModules.delete(path);
          reconcileViteHmrHostProblem();
        });
      }
      return;
    }
    if (viteHmrCycleActive) {
      lastCycleViteErrorDetail = error;
      return;
    }
    tracker.reportProblemNearRender("console-vite-error", error.message, {
      stack: error.stack,
      filename: error.filename
    });
  };
  console.error = patchedConsoleError;
  return () => {
    observer.disconnect();
    stopClearedWatch();
    hotEventsCancelled = true;
    subscribedHot?.off?.("vite:beforeUpdate", onViteBeforeUpdate);
    subscribedHot?.off?.("vite:afterUpdate", onViteAfterUpdate);
    subscribedHot?.off?.("vite:error", onViteError);
    if (console.error === patchedConsoleError) {
      console.error = originalConsoleError;
    }
  };
}
function installCurrentProblemInspector(tracker) {
  const maybeWindow = window;
  const previous = maybeWindow.__tempoErrorTrackingInspectCurrentProblems;
  const inspect = (options = {}) => mergeCurrentProblemSnapshots(
    tracker.inspectCurrentProblems(options),
    previous?.(options) ?? null
  );
  maybeWindow.__tempoErrorTrackingInspectCurrentProblems = inspect;
  return () => {
    if (maybeWindow.__tempoErrorTrackingInspectCurrentProblems === inspect) {
      maybeWindow.__tempoErrorTrackingInspectCurrentProblems = previous;
    }
  };
}
function readViteHot() {
  const explicitHot = window.__tempoCanvasHostViteHot;
  if (explicitHot) return explicitHot;
  try {
    return eval("import.meta && import.meta.hot");
  } catch {
    return void 0;
  }
}
function readCurrentViteOverlay() {
  if (typeof document === "undefined") return void 0;
  const overlay = document.querySelector("vite-error-overlay");
  if (!overlay) return void 0;
  const { message, stack, filename } = readViteOverlayContent(overlay);
  return createProblem("vite-overlay", message, { stack, filename });
}
function readViteOverlayContent(overlay) {
  const root = overlay.shadowRoot ?? overlay;
  const text = (selector) => {
    const value = root.querySelector(selector)?.textContent?.trim();
    return value ? value : void 0;
  };
  const message = text(".message-body") ?? text(".message") ?? overlay.textContent?.trim() ?? "Unknown Vite error";
  const stack = [text(".stack"), text(".plugin"), text(".frame")].filter(Boolean).join("\n");
  return {
    message,
    stack: stack || void 0,
    filename: text(".file")
  };
}
function readViteError(value) {
  if (value instanceof Error) {
    return {
      message: value.message,
      stack: value.stack,
      filename: typeof value.id === "string" ? value.id : void 0
    };
  }
  if (typeof value === "string") return { message: value };
  if (typeof value !== "object" || value === null) return void 0;
  const record = value;
  const nested = typeof record.err === "object" && record.err !== null ? record.err : record;
  const message = typeof nested.message === "string" ? nested.message : typeof record.message === "string" ? record.message : void 0;
  if (!message) return void 0;
  return {
    message,
    stack: typeof nested.stack === "string" ? nested.stack : typeof record.stack === "string" ? record.stack : void 0,
    filename: typeof nested.filename === "string" ? nested.filename : typeof nested.file === "string" ? nested.file : typeof nested.id === "string" ? nested.id : typeof record.id === "string" ? record.id : void 0
  };
}
function readViteConsoleError(first, second) {
  if (first === "[vite]") return readViteError(second);
  if (typeof first !== "string" || !first.startsWith("[vite]")) {
    return void 0;
  }
  const detail = readViteError(second);
  const message = first.replace(/^\[vite\]\s*/, "") || detail?.message;
  return message ? {
    message,
    stack: detail?.stack,
    filename: detail?.filename
  } : void 0;
}
function readUpdatePaths(payload) {
  const updates = typeof payload === "object" && payload !== null && Array.isArray(payload.updates) ? payload.updates : [];
  const paths = /* @__PURE__ */ new Set();
  const collectPath = (value) => {
    if (typeof value === "string") paths.add(value);
  };
  for (const update of updates) {
    collectPath(update.path);
    collectPath(update.acceptedPath);
    collectPath(update.id);
  }
  return [...paths];
}
async function verifyModuleStillExists(path) {
  if (typeof fetch !== "function") return true;
  try {
    const response = await fetch(path, { cache: "no-store" });
    const contentType = response.headers.get("content-type") || "";
    return !contentType.includes("text/html");
  } catch {
    return true;
  }
}
function mergeCurrentProblemSnapshots(primary, fallback) {
  if (!fallback) return primary;
  const currentRenderFailures = [...primary.currentRenderFailures];
  const seen = new Set(
    currentRenderFailures.map(
      (entry) => `${entry.renderId ?? ""}:${entry.failure.sourceKind}:${entry.failure.message}`
    )
  );
  for (const entry of fallback.currentRenderFailures) {
    const key = `${entry.renderId ?? ""}:${entry.failure.sourceKind}:${entry.failure.message}`;
    if (seen.has(key)) continue;
    seen.add(key);
    currentRenderFailures.push(entry);
  }
  const currentCanvasFailures = [...primary.currentCanvasFailures ?? []];
  const seenCanvas = new Set(
    currentCanvasFailures.map(
      (entry) => `${entry.canvasPath}:${entry.failure.sourceKind}:${entry.failure.message}`
    )
  );
  for (const entry of fallback.currentCanvasFailures ?? []) {
    const key = `${entry.canvasPath}:${entry.failure.sourceKind}:${entry.failure.message}`;
    if (seenCanvas.has(key)) continue;
    seenCanvas.add(key);
    currentCanvasFailures.push(entry);
  }
  return {
    currentHostFailure: primary.currentHostFailure ?? fallback.currentHostFailure,
    currentRenderFailures,
    currentCanvasFailures,
    observedAt: Math.max(primary.observedAt, fallback.observedAt)
  };
}

// canvas/use-canvas-host-postmessage.ts
function postToMessageSource(source, response, targetOrigin) {
  const origin = safePostMessageTargetOrigin(targetOrigin);
  if (source && "postMessage" in source && source !== window) {
    source.postMessage(response, origin);
  } else {
    window.postMessage(response, origin);
  }
}
function safePostMessageTargetOrigin(targetOrigin) {
  if (targetOrigin && targetOrigin !== "null" && targetOrigin !== "file://") {
    return targetOrigin;
  }
  const currentOrigin = window.location.origin;
  return currentOrigin && currentOrigin !== "null" ? currentOrigin : "*";
}
function useCanvasHostPostMessage() {
  const [renderEntriesById, setRenderEntriesById] = useState(
    () => /* @__PURE__ */ new Map()
  );
  const [watchRequestsById, setWatchRequestsById] = useState(
    () => /* @__PURE__ */ new Map()
  );
  const [renderCommitVersionById, setRenderCommitVersionById] = useState(
    () => /* @__PURE__ */ new Map()
  );
  const renderEntriesRef = useRef(renderEntriesById);
  const watchRequestsRef = useRef(watchRequestsById);
  const maxCachedRef = useRef(20);
  const renderedRenderIdsRef = useRef(/* @__PURE__ */ new Set());
  const pendingRenderAcksRef = useRef(/* @__PURE__ */ new Map());
  const watchesRef = useRef(/* @__PURE__ */ new Map());
  const problemWatchersRef = useRef(/* @__PURE__ */ new Map());
  const problemTrackerRef = useRef(createCanvasHostProblemTracker());
  const renderCommitTimesRef = useRef(/* @__PURE__ */ new Map());
  const sourceAnnotatedCommitTimesRef = useRef(/* @__PURE__ */ new Map());
  const buildProblemSnapshot = useCallback2(
    () => problemTrackerRef.current.snapshot(),
    []
  );
  const recordSourceAnnotatedCommit = useCallback2(
    (renderId, observedAt = Date.now()) => {
      if (renderSlotHasSourceAnnotations(renderId)) {
        sourceAnnotatedCommitTimesRef.current.set(renderId, observedAt);
      }
    },
    []
  );
  const publishProblemSnapshot = useCallback2(() => {
    const payload = buildProblemSnapshot();
    for (const watcher of problemWatchersRef.current.values()) {
      postToMessageSource(
        watcher.source,
        {
          type: "render-pool:problem-snapshot",
          requestId: watcher.requestId,
          watchId: watcher.watchId,
          payload
        },
        watcher.targetOrigin
      );
    }
  }, [buildProblemSnapshot]);
  const setEntries = useCallback2(
    (update) => {
      setRenderEntriesById((entries) => {
        const next = update(entries);
        renderEntriesRef.current = next;
        for (const renderId of [...renderedRenderIdsRef.current]) {
          if (!next.has(renderId)) {
            renderedRenderIdsRef.current.delete(renderId);
            renderCommitTimesRef.current.delete(renderId);
            sourceAnnotatedCommitTimesRef.current.delete(renderId);
          }
        }
        return next;
      });
    },
    []
  );
  const waitForRender = useCallback2(
    (request) => {
      const current = renderEntriesRef.current.get(request.renderId);
      const alreadyRendered = current?.status === "active" && renderedRenderIdsRef.current.has(request.renderId) && renderRequestMatchesEntry(current, request);
      setEntries((entries) => acquireRender(entries, request));
      if (alreadyRendered) {
        return Promise.resolve();
      }
      return new Promise((resolve, reject) => {
        const pendingAcks = pendingRenderAcksRef.current.get(request.renderId) ?? [];
        pendingAcks.push({ resolve, reject });
        pendingRenderAcksRef.current.set(request.renderId, pendingAcks);
      });
    },
    [setEntries]
  );
  const releaseRenderLease = useCallback2(
    (renderId) => {
      setEntries(
        (entries) => releaseRender(entries, renderId, maxCachedRef.current)
      );
    },
    [setEntries]
  );
  const stopRuntimeWatch = useCallback2(async (watchId) => {
    const existingWatch = watchesRef.current.get(watchId);
    if (!existingWatch) return;
    watchesRef.current.delete(watchId);
    await existingWatch.watch.stop();
  }, []);
  const removeWatchRequest = useCallback2(
    async (watchId) => {
      const existingRequest = watchRequestsRef.current.get(watchId);
      try {
        await stopRuntimeWatch(watchId);
      } finally {
        if (!existingRequest) return;
        setWatchRequestsById((requests) => {
          const next = new Map(requests);
          next.delete(watchId);
          watchRequestsRef.current = next;
          return next;
        });
        releaseRenderLease(existingRequest.request.renderId);
      }
    },
    [releaseRenderLease, stopRuntimeWatch]
  );
  const buildState = useCallback2(
    () => buildRenderStatePayload(renderEntriesRef.current, maxCachedRef.current),
    []
  );
  const buildHealthSnapshot = useCallback2((renderId) => {
    const slot = renderId ? document.querySelector(buildRenderContentSelector(renderId)) : null;
    const observedSourceFiles = getObservedSourceFiles2(slot);
    return {
      alive: true,
      reactConnected: isReactConnected() || renderCommitTimesRef.current.size > 0 || renderEntriesRef.current.size > 0,
      documentReadyState: document.readyState,
      hasViteOverlayElement: document.querySelector("vite-error-overlay") !== null,
      lastHostRenderCommitAt: latestTimestamp(renderCommitTimesRef.current),
      lastClaimRenderCommitAt: renderCommitTimesRef.current.get(renderId) ?? null,
      lastHostSourceAnnotatedCommitAt: latestTimestamp(
        sourceAnnotatedCommitTimesRef.current
      ),
      lastClaimSourceAnnotatedCommitAt: sourceAnnotatedCommitTimesRef.current.get(renderId) ?? null,
      renderSlotExists: slot !== null,
      renderSlotHasSourceAnnotatedContent: observedSourceFiles.length > 0,
      renderSlotObservedSourceFiles: observedSourceFiles,
      claimErrorBoundaryTripped: problemTrackerRef.current.hasComponentProblem(renderId),
      observedAt: Date.now()
    };
  }, []);
  const markStoryboardRendered = useCallback2(
    (storyboardId) => {
      const committedAt = Date.now();
      renderedRenderIdsRef.current.add(storyboardId);
      renderCommitTimesRef.current.set(storyboardId, committedAt);
      recordSourceAnnotatedCommit(storyboardId, committedAt);
      setRenderCommitVersionById((versions) => {
        const next = new Map(versions);
        next.set(storyboardId, (next.get(storyboardId) ?? 0) + 1);
        return next;
      });
      problemTrackerRef.current.clearRecoveredComponentProblemsAfterRender(
        storyboardId
      );
      const pendingAcks = pendingRenderAcksRef.current.get(storyboardId);
      if (!pendingAcks) return;
      pendingRenderAcksRef.current.delete(storyboardId);
      for (const ack of pendingAcks) {
        ack.resolve();
      }
    },
    [recordSourceAnnotatedCommit]
  );
  const reportStoryboardError = useCallback2(
    (storyboardId, error) => {
      renderedRenderIdsRef.current.delete(storyboardId);
      problemTrackerRef.current.reportComponentProblem(
        storyboardId,
        "render-error-boundary",
        error.message,
        {
          stack: error.stack,
          componentStack: error.componentStack
        }
      );
      const pendingAcks = pendingRenderAcksRef.current.get(storyboardId);
      if (pendingAcks) {
        pendingRenderAcksRef.current.delete(storyboardId);
        const renderError = new Error(error.message);
        for (const ack of pendingAcks) {
          ack.reject(renderError);
        }
      }
    },
    []
  );
  const clearStoryboardError = useCallback2(
    (storyboardId) => {
      problemTrackerRef.current.clearComponentProblem(storyboardId, [
        "render-error-boundary"
      ]);
    },
    []
  );
  useEffect(() => {
    if (typeof window === "undefined") return;
    let cancelled = false;
    const activeWatchIds = new Set(watchRequestsById.keys());
    for (const watchId of watchesRef.current.keys()) {
      if (!activeWatchIds.has(watchId)) {
        void stopRuntimeWatch(watchId);
      }
    }
    for (const request of watchRequestsById.values()) {
      const renderEntry = renderEntriesById.get(request.request.renderId);
      const renderReady = renderEntry?.status === "active" && renderedRenderIdsRef.current.has(request.request.renderId) && renderRequestMatchesEntry(renderEntry, request.request);
      if (!renderReady) continue;
      const existing = watchesRef.current.get(request.watchId);
      if (existing?.bindingKey === request.bindingKey) continue;
      void (async () => {
        if (existing) {
          await stopRuntimeWatch(request.watchId);
        }
        if (cancelled) return;
        const onUpdate = (capture) => {
          recordSourceAnnotatedCommit(request.request.renderId);
          postToMessageSource(
            request.source,
            {
              type: "render-pool:watch-update",
              requestId: request.requestId,
              watchId: request.watchId,
              command: request.commandType,
              payload: { capture }
            },
            request.targetOrigin
          );
        };
        const onError = (error) => {
          postToMessageSource(
            request.source,
            {
              type: "render-pool:error",
              requestId: request.requestId,
              watchId: request.watchId,
              command: request.commandType,
              error: error.message
            },
            request.targetOrigin
          );
        };
        const common = {
          onUpdate,
          onError,
          debounceDelay: readOptionalNumber(request.payload, "debounceDelay"),
          maxWait: readOptionalNumber(request.payload, "maxWait"),
          timeout: readOptionalNumber(request.payload, "timeout")
        };
        const watch = request.commandType === "render-pool:watch-html" ? await watchStoryboardHtml(request.request.renderId, common) : await watchStoryboardCombined(request.request.renderId, {
          ...common,
          fiberSelector: typeof request.payload.fiberSelector === "string" ? request.payload.fiberSelector : void 0,
          fiberTags: Array.isArray(request.payload.fiberTags) ? request.payload.fiberTags : void 0,
          knownSourceVersion: typeof request.payload.knownSourceVersion === "object" && request.payload.knownSourceVersion !== null ? request.payload.knownSourceVersion : void 0
        });
        if (cancelled || !watchRequestsRef.current.has(request.watchId)) {
          await watch.stop();
          return;
        }
        watchesRef.current.set(request.watchId, {
          renderId: request.request.renderId,
          bindingKey: request.bindingKey,
          watch
        });
        if (request.paused) {
          if (watch.setPaused) watch.setPaused(true);
          else watch.pause?.();
        }
      })().catch((caught) => {
        postToMessageSource(
          request.source,
          {
            type: "render-pool:error",
            requestId: request.requestId,
            watchId: request.watchId,
            command: request.commandType,
            error: caught instanceof Error ? caught.message : String(caught)
          },
          request.targetOrigin
        );
      });
    }
    return () => {
      cancelled = true;
    };
  }, [
    recordSourceAnnotatedCommit,
    renderCommitVersionById,
    renderEntriesById,
    stopRuntimeWatch,
    watchRequestsById
  ]);
  useEffect(() => {
    if (typeof window === "undefined") return;
    const waitForRenderLease = async (request) => {
      try {
        await waitForRender(request);
      } catch (caught) {
        releaseRenderLease(request.renderId);
        throw caught;
      }
    };
    const withRenderLease = async (request, run) => {
      await waitForRenderLease(request);
      try {
        return await run();
      } finally {
        releaseRenderLease(request.renderId);
      }
    };
    const handleCommand = (message, source, targetOrigin) => {
      const command = readCanvasHostCommand(message);
      if (!command) return;
      const payload = command.payload;
      const post = (response) => postToMessageSource(source, response, targetOrigin);
      const ack = () => post({
        type: "render-pool:ack",
        requestId: command.requestId,
        command: command.type
      });
      const error = (message2) => post({
        type: "render-pool:error",
        requestId: command.requestId,
        command: command.type,
        error: message2
      });
      const postAsyncError = (caught) => {
        error(caught instanceof Error ? caught.message : String(caught));
      };
      const postCaptureResult = (capture) => post({
        type: "render-pool:capture-result",
        requestId: command.requestId,
        command: command.type,
        payload: { capture }
      });
      try {
        if (command.type === "render-pool:render") {
          const request = readRenderRequest(command.type, payload);
          void waitForRenderLease(request).then(ack).catch(postAsyncError);
          return;
        }
        if (command.type === "render-pool:capture-image") {
          const request = readRenderRequest(command.type, payload);
          void withRenderLease(
            request,
            () => captureStoryboardImage(request.renderId, {
              overlaySelector: typeof payload.overlaySelector === "string" ? payload.overlaySelector : void 0,
              format: payload.format === "jpeg" || payload.format === "png" ? payload.format : void 0,
              quality: readOptionalNumber(payload, "quality"),
              scale: readOptionalNumber(payload, "scale"),
              outputPath: typeof payload.outputPath === "string" ? payload.outputPath : void 0,
              forceRefresh: payload.forceRefresh === true,
              portalSelectors: Array.isArray(payload.portalSelectors) ? payload.portalSelectors.filter(
                (selector) => typeof selector === "string"
              ) : void 0
            })
          ).then((capture) => {
            recordSourceAnnotatedCommit(request.renderId);
            postCaptureResult(capture);
          }).catch(postAsyncError);
          return;
        }
        if (command.type === "render-pool:capture-html") {
          const request = readRenderRequest(command.type, payload);
          void withRenderLease(
            request,
            () => captureStoryboardHtml(request.renderId, {
              timeout: readOptionalNumber(payload, "timeout")
            })
          ).then((capture) => {
            recordSourceAnnotatedCommit(request.renderId);
            postCaptureResult(capture);
          }).catch(postAsyncError);
          return;
        }
        if (command.type === "render-pool:capture-static-html") {
          const request = readRenderRequest(command.type, payload);
          void withRenderLease(
            request,
            () => captureStoryboardStaticHtml(request.renderId, {
              timeout: readOptionalNumber(payload, "timeout")
            })
          ).then((capture) => {
            recordSourceAnnotatedCommit(request.renderId);
            postCaptureResult(capture);
          }).catch(postAsyncError);
          return;
        }
        if (command.type === "render-pool:capture-combined") {
          const request = readRenderRequest(command.type, payload);
          void withRenderLease(
            request,
            () => captureStoryboardCombined(request.renderId, {
              fiberSelector: typeof payload.fiberSelector === "string" ? payload.fiberSelector : void 0,
              fiberTags: Array.isArray(payload.fiberTags) ? payload.fiberTags : void 0,
              timeout: readOptionalNumber(payload, "timeout")
            })
          ).then((capture) => {
            recordSourceAnnotatedCommit(request.renderId);
            postCaptureResult(capture);
          }).catch(postAsyncError);
          return;
        }
        if (command.type === "render-pool:watch-html" || command.type === "render-pool:watch-combined") {
          const request = readRenderRequest(command.type, payload);
          const watchId = readWatchId(command.type, payload);
          void removeWatchRequest(watchId).then(() => {
            setEntries((entries) => acquireRender(entries, request));
            setWatchRequestsById((requests) => {
              const next = new Map(requests);
              next.set(watchId, {
                commandType: command.type,
                requestId: command.requestId,
                watchId,
                source,
                targetOrigin,
                request,
                payload,
                paused: false,
                bindingKey: buildWatchBindingKey(command.type, payload)
              });
              watchRequestsRef.current = next;
              return next;
            });
            ack();
          }).catch(postAsyncError);
          return;
        }
        if (command.type === "render-pool:unwatch") {
          const watchId = readWatchId(command.type, payload);
          void removeWatchRequest(watchId).then(ack).catch(postAsyncError);
          return;
        }
        if (command.type === "render-pool:set-watch-paused") {
          const watchId = readWatchId(command.type, payload);
          const paused = payload.paused === true;
          const watch = watchesRef.current.get(watchId)?.watch;
          watch?.setPaused?.(paused);
          if (!watch?.setPaused) {
            if (paused) watch?.pause?.();
            else watch?.resume?.();
          }
          setWatchRequestsById((requests) => {
            const current = requests.get(watchId);
            if (!current || current.paused === paused) return requests;
            const next = new Map(requests);
            next.set(watchId, { ...current, paused });
            watchRequestsRef.current = next;
            return next;
          });
          ack();
          return;
        }
        if (command.type === "render-pool:unrender") {
          const target = requireRenderTarget(command.type, payload);
          releaseRenderLease(target.renderId);
          ack();
          return;
        }
        if (command.type === "render-pool:list") {
          post({
            type: "render-pool:state",
            requestId: command.requestId,
            payload: buildState()
          });
          return;
        }
        if (command.type === "render-pool:set-cache-limit") {
          const maxCached = typeof payload.maxCached === "number" && Number.isInteger(payload.maxCached) ? Math.max(0, payload.maxCached) : maxCachedRef.current;
          maxCachedRef.current = maxCached;
          setEntries((entries) => evictCachedRenders(entries, maxCached));
          ack();
          return;
        }
        if (command.type === "render-pool:isolate") {
          const target = resolveRenderTarget(payload);
          document.querySelector("[data-render-isolation]")?.remove();
          if (target?.renderId) {
            const style = document.createElement("style");
            style.setAttribute("data-render-isolation", "");
            const escapedId = typeof CSS !== "undefined" && typeof CSS.escape === "function" ? CSS.escape(target.renderId) : target.renderId;
            style.textContent = '[data-render-id]:not([data-render-id="' + escapedId + '"]) { opacity: 0 !important; pointer-events: none !important; }';
            document.head.appendChild(style);
          }
          ack();
          return;
        }
        if (command.type === "render-pool:query-health") {
          const target = resolveRenderTarget(payload);
          post({
            type: "render-pool:health-snapshot",
            requestId: command.requestId,
            snapshot: buildHealthSnapshot(target?.renderId ?? "")
          });
          return;
        }
        if (command.type === "render-pool:query-problems") {
          post({
            type: "render-pool:problem-snapshot",
            requestId: command.requestId,
            payload: buildProblemSnapshot()
          });
          return;
        }
        if (command.type === "render-pool:watch-problems") {
          const watchId = typeof payload.watchId === "string" && payload.watchId.trim() ? payload.watchId : String(command.requestId ?? "default");
          problemWatchersRef.current.set(watchId, {
            requestId: command.requestId,
            watchId,
            source,
            targetOrigin
          });
          ack();
          post({
            type: "render-pool:problem-snapshot",
            requestId: command.requestId,
            watchId,
            payload: buildProblemSnapshot()
          });
          return;
        }
        if (command.type === "render-pool:unwatch-problems") {
          const watchId = typeof payload.watchId === "string" && payload.watchId.trim() ? payload.watchId : String(command.requestId ?? "default");
          problemWatchersRef.current.delete(watchId);
          ack();
        }
      } catch (caught) {
        postAsyncError(caught);
      }
    };
    const onMessage = (event) => handleCommand(event.data, event.source, event.origin);
    const onBridgeCommand = (event) => handleCommand(event instanceof CustomEvent ? event.detail : void 0);
    window.addEventListener("message", onMessage);
    window.addEventListener("__render-pool:command", onBridgeCommand);
    const stopProblemPublisher = problemTrackerRef.current.watch(
      publishProblemSnapshot
    );
    const stopRuntimeProblemTracking = installCanvasHostRuntimeProblemTracking(
      problemTrackerRef.current
    );
    window.postMessage({ type: "render-pool:ready" }, "*");
    return () => {
      window.removeEventListener("message", onMessage);
      window.removeEventListener("__render-pool:command", onBridgeCommand);
      stopRuntimeProblemTracking();
      stopProblemPublisher();
      for (const { watch } of watchesRef.current.values()) {
        void watch.stop();
      }
      watchesRef.current.clear();
      problemWatchersRef.current.clear();
    };
  }, [
    buildHealthSnapshot,
    buildProblemSnapshot,
    buildState,
    publishProblemSnapshot,
    recordSourceAnnotatedCommit,
    releaseRenderLease,
    removeWatchRequest,
    setEntries,
    waitForRender
  ]);
  return {
    renderStatusById: useMemo2(
      () => buildRenderStatusMap(renderEntriesById),
      [renderEntriesById]
    ),
    renderSizeById: useMemo2(
      () => buildRenderSizeMap(renderEntriesById),
      [renderEntriesById]
    ),
    markStoryboardRendered,
    reportStoryboardError,
    clearStoryboardError
  };
}
function latestTimestamp(values) {
  let latest = null;
  for (const value of values.values()) {
    latest = latest === null ? value : Math.max(latest, value);
  }
  return latest;
}
function isReactConnected() {
  const hook = window.__REACT_DEVTOOLS_GLOBAL_HOOK__;
  return (hook?.renderers?.size ?? 0) > 0;
}
function buildWatchBindingKey(commandType, payload) {
  try {
    return JSON.stringify({ commandType, payload });
  } catch {
    return `${commandType}:${Date.now()}`;
  }
}
var sourceAnnotationSelector2 = "[data-tempo-source-file], [data-tempo-source-loc], [data-tempo-filepath], [data-tempo-html-filepath]";
function renderSlotHasSourceAnnotations(renderId) {
  const slot = typeof document === "undefined" ? null : document.querySelector(buildRenderContentSelector(renderId));
  return getObservedSourceFiles2(slot).length > 0;
}
function getObservedSourceFiles2(slot) {
  if (!slot) return [];
  return [
    ...new Set(
      [...slot.querySelectorAll(sourceAnnotationSelector2)].flatMap((element) => [
        element.getAttribute("data-tempo-source-file"),
        element.getAttribute("data-tempo-filepath"),
        element.getAttribute("data-tempo-html-filepath"),
        element.getAttribute("data-tempo-source-loc")
      ]).filter((file) => Boolean(file))
    )
  ];
}

// canvas/canvas-host.tsx
import { jsx as jsx2 } from "react/jsx-runtime";
function readDirectUrlDimension(value) {
  const match = value.trim().match(/^([0-9]*\.?[0-9]+)(px|vw|vh)?$/);
  if (!match) {
    throw new Error(
      "Direct URL dimensions must be positive finite numbers, optionally with px, vw, or vh units"
    );
  }
  const numberPart = match[1] ?? "";
  const unit = match[2] ?? "";
  const numberValue = Number(numberPart);
  if (!numberPart || !Number.isFinite(numberValue) || numberValue <= 0) {
    throw new Error(
      "Direct URL dimensions must be positive finite numbers, optionally with px, vw, or vh units"
    );
  }
  return unit && unit !== "px" ? `${numberValue}${unit}` : numberValue;
}
function readDirectUrlSize(params) {
  const width = params.get("width")?.trim();
  const height = params.get("height")?.trim();
  if (!width && !height) return void 0;
  if (!width || !height) {
    throw new Error("Direct URL width and height must be provided together");
  }
  return {
    width: readDirectUrlDimension(width),
    height: readDirectUrlDimension(height),
    revision: 1
  };
}
function readDirectUrlRender() {
  if (typeof window === "undefined") return null;
  const params = new URLSearchParams(window.location.search);
  const canvasPath = params.get("canvasPath")?.trim() ?? params.get("pageFile")?.trim() ?? params.get("importPath")?.trim() ?? "";
  const storyboardId = params.get("storyboardId")?.trim() ?? params.get("exportName")?.trim() ?? params.get("importName")?.trim() ?? "";
  if (!canvasPath || !storyboardId) return null;
  return {
    renderId: getCanvasStoryboardRenderId(canvasPath, storyboardId),
    size: readDirectUrlSize(params)
  };
}
function CanvasHost({
  children,
  activeStoryboardIds,
  isStoryboardActive
}) {
  const {
    renderStatusById: renderStatusById2,
    renderSizeById,
    markStoryboardRendered,
    reportStoryboardError,
    clearStoryboardError
  } = useCanvasHostPostMessage();
  const directUrlSearch = typeof window === "undefined" ? "" : window.location.search;
  const directUrlRender = useMemo3(readDirectUrlRender, [directUrlSearch]);
  const propActiveIds = useMemo3(
    () => new Set(activeStoryboardIds ?? []),
    [activeStoryboardIds]
  );
  const runtime = useMemo3(
    () => ({
      getStoryboardStatus(storyboardId) {
        const normalizedId = normalizeId(storyboardId, "Storyboard id");
        if (propActiveIds.has(normalizedId) || directUrlRender?.renderId === normalizedId || (isStoryboardActive ? isStoryboardActive(normalizedId) : false)) {
          return "active";
        }
        return renderStatusById2.get(normalizedId) ?? "unmounted";
      },
      getStoryboardSize(storyboardId) {
        const normalizedId = normalizeId(storyboardId, "Storyboard id");
        if (directUrlRender?.renderId === normalizedId) {
          return directUrlRender.size;
        }
        return renderSizeById.get(normalizedId);
      },
      getStoryboardPlacement(storyboardId) {
        const normalizedId = normalizeId(storyboardId, "Storyboard id");
        return directUrlRender?.renderId === normalizedId ? { left: 0, top: 0 } : void 0;
      },
      markStoryboardRendered,
      reportStoryboardError,
      clearStoryboardError
    }),
    [
      clearStoryboardError,
      directUrlRender,
      isStoryboardActive,
      markStoryboardRendered,
      propActiveIds,
      reportStoryboardError,
      renderSizeById,
      renderStatusById2
    ]
  );
  return /* @__PURE__ */ jsx2(CanvasHostContext.Provider, { value: runtime, children });
}

// canvas/storyboard.tsx
import {
  Component,
  Suspense,
  useEffect as useEffect2,
  useRef as useRef2
} from "react";
import { Fragment, jsx as jsx3, jsxs } from "react/jsx-runtime";
function Storyboard({
  id,
  name,
  component,
  layout,
  props
}) {
  const Component2 = component;
  return /* @__PURE__ */ jsx3(StoryboardSlot, { id, name, layout, kind: "component", children: /* @__PURE__ */ jsx3(Component2, { ...props ?? {} }) });
}
function LegacyStoryboardV1({
  id,
  storyboard
}) {
  return /* @__PURE__ */ jsx3(
    StoryboardSlot,
    {
      id,
      name: storyboard.name,
      layout: storyboard.layout,
      kind: "component",
      children: /* @__PURE__ */ jsx3(LegacyStoryboardV1Render, { render: storyboard.render })
    }
  );
}
function LegacyStoryboardV1Render({
  render
}) {
  return /* @__PURE__ */ jsx3(Fragment, { children: render() });
}
function RouteStoryboard({
  id,
  name,
  route,
  layout
}) {
  return /* @__PURE__ */ jsx3(
    StoryboardSlot,
    {
      id,
      name,
      layout,
      kind: "route",
      route
    }
  );
}
function HtmlStoryboard({
  id,
  name,
  filepath,
  src,
  layout
}) {
  return /* @__PURE__ */ jsx3(
    StoryboardSlot,
    {
      id,
      name,
      layout,
      kind: "html",
      filepath,
      src
    }
  );
}
function StoryboardSlot({
  children,
  id,
  kind,
  layout,
  name,
  route,
  filepath,
  src
}) {
  const renderId = useCanvasStoryboardRenderId(id);
  const canvas = useCanvasIdentity();
  const status = useStoryboardStatus(id);
  const renderSize = useStoryboardSize(id);
  const placement = useStoryboardPlacement(id);
  const markRendered = useStoryboardRendered(id);
  const { reportError, clearError } = useStoryboardErrorReporter(id);
  const renderWidth = renderSize?.width;
  const renderHeight = renderSize?.height;
  const intrinsicSize = renderSize?.intrinsicSize ?? (kind === "component" && layout.intrinsicSizing === "root-element");
  const resetKey = `${renderId}:${renderSize?.revision ?? 0}`;
  const contentRef = useRef2(null);
  useEffect2(() => {
    if (!contentRef.current || status === "unmounted") return;
    return trackStoryboardPortals(renderId, contentRef.current, status);
  }, [renderId, status]);
  useEffect2(() => {
    if (status === "unmounted") return;
    syncStoryboardPortals(renderId, status);
  }, [renderId, status]);
  if (status === "unmounted") return null;
  return /* @__PURE__ */ jsx3(
    "div",
    {
      "data-render-id": renderId,
      "data-render-status": status,
      "data-tempo-storyboard-render-id": renderId,
      "data-tempo-storyboard-id": id,
      "data-tempo-canvas-path": canvas?.path,
      "data-tempo-storyboard-name": name,
      "data-tempo-storyboard-kind": kind,
      "data-tempo-storyboard-route": route,
      "data-tempo-storyboard-filepath": filepath,
      "data-tempo-storyboard-src": src,
      style: {
        position: "absolute",
        left: placement?.left ?? layout.x,
        top: placement?.top ?? layout.y,
        width: intrinsicSize ? "max-content" : renderWidth ?? layout.width,
        height: intrinsicSize ? "max-content" : renderHeight ?? layout.height,
        minWidth: intrinsicSize ? renderWidth : void 0,
        minHeight: intrinsicSize ? renderHeight : void 0,
        zIndex: layout.zIndex,
        visibility: status === "active" ? "visible" : "hidden",
        pointerEvents: status === "active" ? "auto" : "none"
      },
      children: /* @__PURE__ */ jsx3(
        "div",
        {
          "data-render-content": renderId,
          ref: contentRef,
          style: { width: "100%", height: "100%" },
          children: /* @__PURE__ */ jsx3(
            StoryboardErrorBoundary,
            {
              onError: reportError,
              onRecover: clearError,
              resetKey,
              children: /* @__PURE__ */ jsxs(Suspense, { fallback: null, children: [
                children,
                /* @__PURE__ */ jsx3(
                  StoryboardReadySentinel,
                  {
                    active: status === "active",
                    clearError,
                    markRendered
                  }
                )
              ] })
            }
          )
        }
      )
    }
  );
}
var renderErrorRetryDelayMs = 1e3;
var StoryboardErrorBoundary = class extends Component {
  constructor() {
    super(...arguments);
    this.retryTimer = null;
    this.state = {
      error: null,
      componentStack: void 0,
      resetKey: this.props.resetKey
    };
  }
  static getDerivedStateFromError(error) {
    return { error };
  }
  static getDerivedStateFromProps(props, state) {
    if (props.resetKey !== state.resetKey) {
      return { error: null, componentStack: void 0, resetKey: props.resetKey };
    }
    return null;
  }
  componentDidCatch(error, info) {
    this.setState({ componentStack: info.componentStack ?? void 0 });
    this.props.onError({
      message: error.message,
      stack: error.stack,
      componentStack: info.componentStack ?? void 0
    });
    this.scheduleRetry();
  }
  componentDidUpdate(_prevProps, prevState) {
    if (this.state.error) {
      if (!this.retryTimer) this.scheduleRetry();
      return;
    }
    this.clearRetryTimer();
    if (prevState.error && !this.state.error) {
      this.props.onRecover();
    }
  }
  componentWillUnmount() {
    this.clearRetryTimer();
  }
  scheduleRetry() {
    this.clearRetryTimer();
    this.retryTimer = setTimeout(() => {
      this.retryTimer = null;
      this.setState(
        (state) => state.error ? { error: null, componentStack: void 0, resetKey: state.resetKey } : null
      );
    }, renderErrorRetryDelayMs);
  }
  clearRetryTimer() {
    if (!this.retryTimer) return;
    clearTimeout(this.retryTimer);
    this.retryTimer = null;
  }
  render() {
    if (this.state.error) {
      return /* @__PURE__ */ jsx3(
        "div",
        {
          "data-render-error": "true",
          "data-render-error-message": this.state.error.message,
          style: { width: "100%", height: "100%" }
        }
      );
    }
    return this.props.children;
  }
};
function StoryboardReadySentinel({
  active,
  clearError,
  markRendered
}) {
  useEffect2(() => {
    if (!active) return;
    let paintFrame = 0;
    const commitFrame = requestAnimationFrame(() => {
      paintFrame = requestAnimationFrame(() => {
        clearError();
        markRendered();
      });
    });
    return () => {
      cancelAnimationFrame(commitFrame);
      cancelAnimationFrame(paintFrame);
    };
  }, [active, clearError, markRendered]);
  return null;
}
export {
  Canvas,
  CanvasContext,
  CanvasHost,
  HtmlStoryboard,
  LegacyStoryboardV1,
  RouteStoryboard,
  Storyboard,
  StoryboardSlot,
  buildRenderContentFiberSelector,
  buildRenderContentSelector,
  buildRenderSelector,
  captureStoryboardCombined,
  captureStoryboardHtml,
  captureStoryboardImage,
  captureStoryboardStaticHtml,
  getCanvasHostRuntime,
  getCanvasStoryboardRenderId,
  getOptionalCanvasHostRuntime,
  stripRenderContentWrapperAttributes,
  stripRenderContentWrapperFromCombinedCapture,
  useActiveState,
  watchStoryboardCombined,
  watchStoryboardHtml
};
