// tempo-sdk/dist/nextjs/plugin.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// nextjs/plugin.ts
var plugin_exports = {};
__export(plugin_exports, {
  default: () => tempoNextjsPlugin
});
module.exports = __toCommonJS(plugin_exports);

// nextjs.ts
var import_node_fs = require("fs");
var import_node_module = require("module");
var import_node_path = require("path");
var import_meta = {};
var TURBOPACK_RULE_KEY = "*.{tsx,jsx}";
var requireFromHere = (0, import_node_module.createRequire)(
  typeof import_meta.url === "string" && import_meta.url.length > 0 ? import_meta.url : __filename
);
function resolveLoaderPath() {
  for (const candidate of ["./nextjs-loader.cjs", "./dist/nextjs-loader.cjs"]) {
    try {
      return requireFromHere.resolve(candidate);
    } catch {
    }
  }
  throw new Error(
    "tempo-sdk/nextjs: could not resolve nextjs-loader.cjs. Run `pnpm build` in the tempo-sdk package to produce the dist/ output."
  );
}
function detectNextVersion() {
  try {
    const pkg = requireFromHere("next/package.json");
    const m = /^(\d+)\.(\d+)/.exec(pkg.version);
    if (!m) return null;
    return { major: Number(m[1]), minor: Number(m[2]) };
  } catch {
    return null;
  }
}
function useExperimentalTurbo(version) {
  if (!version) return false;
  if (version.major < 15) return true;
  if (version.major === 15 && version.minor < 3) return true;
  return false;
}
function shouldOmitTurbopack(version) {
  if (!version) return false;
  return version.major < 13;
}
function buildTurbopackRule(loaderPath, options) {
  const loaderOptions = {};
  if (options.computeAnnotatedPropsModule !== void 0) {
    loaderOptions.computeAnnotatedPropsModule = options.computeAnnotatedPropsModule;
  }
  if (options.autoInjectRootMeta !== void 0) {
    loaderOptions.autoInjectRootMeta = options.autoInjectRootMeta;
  }
  const condition = options.productionBuild ? { not: "foreign" } : { all: [{ not: "foreign" }, "development"] };
  return {
    condition,
    loaders: [{ loader: loaderPath, options: loaderOptions }]
  };
}
function mergeTurbopackRules(existing, newRule) {
  const existingRule = existing?.rules?.[TURBOPACK_RULE_KEY];
  let mergedRule;
  if (!existingRule) {
    mergedRule = newRule;
  } else if (Array.isArray(existingRule)) {
    mergedRule = [...existingRule, newRule];
  } else {
    mergedRule = [existingRule, newRule];
  }
  return {
    ...existing,
    rules: {
      ...existing?.rules,
      [TURBOPACK_RULE_KEY]: mergedRule
    }
  };
}
function buildWebpackHook(loaderPath, options, userWebpack) {
  const loaderOptions = {};
  if (options.computeAnnotatedPropsModule !== void 0) {
    loaderOptions.computeAnnotatedPropsModule = options.computeAnnotatedPropsModule;
  }
  if (options.autoInjectRootMeta !== void 0) {
    loaderOptions.autoInjectRootMeta = options.autoInjectRootMeta;
  }
  const allowProduction = options.productionBuild === true;
  return (configIn, context) => {
    var _a;
    const config = configIn;
    const ctx = context;
    if (ctx.dev || allowProduction) {
      config.module ?? (config.module = {});
      (_a = config.module).rules ?? (_a.rules = []);
      config.module.rules.push({
        test: /\.(tsx|jsx)$/,
        exclude: /node_modules/,
        enforce: "pre",
        use: [{ loader: loaderPath, options: loaderOptions }]
      });
    }
    if (typeof userWebpack === "function") {
      return userWebpack(config, context);
    }
    return config;
  };
}
function tempoNextjsPlugin(pluginOptions = {}) {
  return (inputConfigArg) => {
    const inputConfig = inputConfigArg ?? {};
    if (!process.env.TEMPO) return inputConfig;
    const loaderPath = _internal.resolveLoaderPath();
    const version = _internal.detectNextVersion();
    const omitTurbopack = _internal.shouldOmitTurbopack(version);
    const useExperimental = _internal.useExperimentalTurbo(version);
    const newRule = buildTurbopackRule(loaderPath, pluginOptions);
    const existingTurbo = useExperimental ? inputConfig.experimental?.turbo : inputConfig.turbopack;
    const mergedTurbo = mergeTurbopackRules(existingTurbo, newRule);
    const webpackHook = buildWebpackHook(
      loaderPath,
      pluginOptions,
      inputConfig.webpack ?? null
    );
    const next = {
      ...inputConfig,
      webpack: webpackHook
    };
    if (!omitTurbopack) {
      if (useExperimental) {
        next.experimental = {
          ...inputConfig.experimental,
          turbo: mergedTurbo
        };
      } else {
        next.turbopack = mergedTurbo;
      }
    }
    return next;
  };
}
var _internal = {
  buildTurbopackRule,
  mergeTurbopackRules,
  buildWebpackHook,
  detectNextVersion,
  useExperimentalTurbo,
  shouldOmitTurbopack,
  resolveLoaderPath
};
module.exports = Object.assign(module.exports.default, module.exports);
