// tempo-sdk/dist/nextjs-loader.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// nextjs-loader.ts
var nextjs_loader_exports = {};
__export(nextjs_loader_exports, {
  _resetLoaderState: () => _resetLoaderState,
  default: () => nextjs_loader_default,
  injectRootMetaIntoAnnotatedSource: () => injectRootMetaIntoAnnotatedSource,
  injectRootMetaIntoSource: () => injectRootMetaIntoSource,
  nextjsLoader: () => nextjsLoader
});
module.exports = __toCommonJS(nextjs_loader_exports);
var import_node_fs2 = require("fs");
var import_node_module = require("module");
var import_node_path2 = require("path");

// ../../modules/annotation/annotate-file.ts
var import_oxc_parser2 = require("oxc-parser");

// ../../modules/file-parser/utils.ts
function getTagName(name) {
  switch (name.type) {
    case "JSXIdentifier":
      return name.name;
    case "JSXNamespacedName":
      return `${name.namespace.name}:${name.name.name}`;
    case "JSXMemberExpression":
      return getJSXMemberExpressionName(name);
  }
}
function getJSXMemberExpressionName(node) {
  const object = node.object.type === "JSXIdentifier" ? node.object.name : getJSXMemberExpressionName(node.object);
  return `${object}.${node.property.name}`;
}

// ../../modules/file-parser/walk.ts
function walk(node, callbacks) {
  if (node === null || node === void 0 || typeof node !== "object") return;
  if (Array.isArray(node)) {
    for (const child of node) {
      walk(child, callbacks);
    }
    return;
  }
  const record = node;
  const type = record.type;
  if (typeof type === "string" && type in callbacks) {
    callbacks[type](node);
  }
  for (const value of Object.values(record)) {
    if (value !== null && typeof value === "object") {
      walk(value, callbacks);
    }
  }
}

// ../../modules/file-parser/scope-analyzer/types.ts
var ScopeFlags = {
  Top: 1 << 0,
  Function: 1 << 1,
  Block: 1 << 2,
  Catch: 1 << 3,
  Class: 1 << 4,
  For: 1 << 5,
  Switch: 1 << 6,
  Module: 1 << 7,
  StrictMode: 1 << 8,
  ClassStaticBlock: 1 << 9,
  TsModuleBlock: 1 << 10,
  TsConditional: 1 << 11,
  Type: 1 << 12,
  Enum: 1 << 13
};

// ../../modules/file-parser/scope-analyzer/scopeAnalyzer.ts
var import_oxc_parser = require("oxc-parser");
var import_keys = __toESM(require("oxc-parser/src-js/generated/visit/keys.js"), 1);

// ../../modules/annotation/annotate-file.ts
function formatPropValue(value) {
  if (value === void 0) return null;
  if (typeof value === "function") return value.toString();
  return JSON.stringify(value);
}
function formatProps(props) {
  return Object.entries(props).map(([key, value]) => {
    const formatted = formatPropValue(value);
    if (formatted === null) return null;
    return ` ${key}={${formatted}}`;
  }).filter((s) => s !== null).join("");
}
function getExistingAttributeNames(element) {
  const names = /* @__PURE__ */ new Set();
  for (const attr of element.openingElement.attributes ?? []) {
    if (attr.type !== "JSXAttribute") continue;
    names.add(getTagName(attr.name));
  }
  return names;
}
function removeExistingProps(props, existingNames) {
  return Object.fromEntries(
    Object.entries(props).filter(([key]) => !existingNames.has(key))
  );
}
var FRAGMENT_TAG_NAMES = /* @__PURE__ */ new Set(["Fragment", "React.Fragment"]);
function annotateFile(source, filename, computeAnnotatedProps) {
  const parseResult = (0, import_oxc_parser2.parseSync)(filename, source);
  const insertions = [];
  walk(parseResult.program, {
    JSXElement(node) {
      const element = node;
      const tagName = getTagName(element.openingElement.name);
      if (FRAGMENT_TAG_NAMES.has(tagName)) return;
      const props = removeExistingProps(
        computeAnnotatedProps({ element, tagName }),
        getExistingAttributeNames(element)
      );
      if (Object.keys(props).length === 0) return;
      const openingElement = element.openingElement;
      let insertPos = openingElement.end - 1;
      if (openingElement.selfClosing) {
        insertPos--;
        while (insertPos > openingElement.start && /\s/.test(source[insertPos - 1])) {
          insertPos--;
        }
      }
      insertions.push({ position: insertPos, text: formatProps(props) });
    }
  });
  insertions.sort((a, b) => a.position - b.position);
  const chunks = [];
  let lastPos = 0;
  for (const { position, text } of insertions) {
    chunks.push(source.slice(lastPos, position));
    chunks.push(text);
    lastPos = position;
  }
  chunks.push(source.slice(lastPos));
  return { annotatedSource: chunks.join(""), elementsAnnotated: insertions.length, parseResult };
}

// ../../modules/annotation/source-version.ts
function sourceVersionHash(content) {
  let djb2 = 5381;
  let sdbm = 0;
  for (let i = 0; i < content.length; i++) {
    const c = content.charCodeAt(i);
    djb2 = (djb2 << 5) + djb2 ^ c;
    sdbm = c + (sdbm << 6) + (sdbm << 16) - sdbm | 0;
  }
  return toHex8(djb2) + toHex8(sdbm);
}
function toHex8(n) {
  return (n >>> 0).toString(16).padStart(8, "0");
}

// nextjs-loader.ts
var import_oxc_parser3 = require("oxc-parser");

// annotation-strategy.ts
var import_node_fs = require("fs");
var import_node_path = require("path");
var HTML_VOID_ELEMENTS = /* @__PURE__ */ new Set([
  "area",
  "base",
  "br",
  "col",
  "embed",
  "hr",
  "img",
  "input",
  "link",
  "meta",
  "param",
  "source",
  "track",
  "wbr"
]);
function isIntrinsicElement(tagName) {
  return tagName.length > 0 && tagName[0] === tagName[0].toLowerCase();
}
function hasJsxAttribute(info, attrName) {
  for (const attr of info.element.openingElement.attributes) {
    const name = attr.name;
    if (name?.name === attrName) return true;
  }
  return false;
}
function defaultComputeAnnotatedProps(info, filepath, modifiedtsMs, sourceVersion) {
  return {
    "data-tempo-filepath": filepath,
    "data-tempo-position": info.element.openingElement.start,
    "data-tempo-modifiedts": modifiedtsMs,
    // Content-hash identity of the file this element was annotated from. The
    // canvas reconcile loop equality-compares it against the write pipeline's
    // computed version to decide whether a capture reflects a given edit.
    "data-tempo-source-version": sourceVersion
  };
}
function computeTempoAnnotatedProps(params) {
  const {
    info,
    filepath,
    modifiedtsMs,
    sourceVersion,
    computeAnnotatedProps,
    elementSupportOverrides
  } = params;
  const userProps = computeAnnotatedProps ? computeAnnotatedProps(info, filepath, modifiedtsMs, sourceVersion) : defaultComputeAnnotatedProps(info, filepath, modifiedtsMs, sourceVersion);
  if (isIntrinsicElement(info.tagName)) {
    return {
      ...userProps,
      "data-tempo-supports-children": hasJsxAttribute(
        info,
        "data-tempo-supports-children"
      ) ? void 0 : !HTML_VOID_ELEMENTS.has(info.tagName)
    };
  }
  const override = elementSupportOverrides?.get(info.tagName);
  if (override) {
    return {
      ...userProps,
      "data-tempo-supports-children": hasJsxAttribute(
        info,
        "data-tempo-supports-children"
      ) ? void 0 : override.supportsChildren
    };
  }
  return userProps;
}

// nextjs-loader.ts
var ROOT_META_NAME = "tempo-project-root";
var cachedCanonicalRoot = null;
var cachedUserComputeProps;
var cachedComputeModulePath = null;
function _resetLoaderState() {
  cachedCanonicalRoot = null;
  cachedUserComputeProps = void 0;
  cachedComputeModulePath = null;
}
function stripQueryAndHash(path) {
  const queryIdx = path.indexOf("?");
  const hashIdx = path.indexOf("#");
  let end = path.length;
  if (queryIdx !== -1) end = Math.min(end, queryIdx);
  if (hashIdx !== -1) end = Math.min(end, hashIdx);
  return path.slice(0, end);
}
function resolveCanonicalRoot(rootContext) {
  try {
    return import_node_fs2.realpathSync.native(rootContext);
  } catch {
    return rootContext;
  }
}
function isInNodeModules(path) {
  return path.includes(`${import_node_path2.sep}node_modules${import_node_path2.sep}`);
}
function findFirstJSXElement(program, tagName) {
  let result = null;
  walk(program, {
    JSXElement(node) {
      if (result) return;
      const element = node;
      if (getTagName(element.openingElement.name) === tagName) {
        result = element;
      }
    }
  });
  return result;
}
function hasRootMetaChild(headElement) {
  for (const child of headElement.children ?? []) {
    if (child.type !== "JSXElement") continue;
    const childTag = getTagName(child.openingElement.name);
    if (childTag !== "meta") continue;
    for (const attr of child.openingElement.attributes ?? []) {
      if (attr.type !== "JSXAttribute") continue;
      if (getTagName(attr.name) !== "name") continue;
      const value = attr.value;
      if (value && value.type === "Literal" && value.value === ROOT_META_NAME) {
        return true;
      }
    }
  }
  return false;
}
function injectFirstChildMarkup(source, element, childMarkup) {
  const opening = element.openingElement;
  if (!opening.selfClosing) {
    const insertPos = opening.end;
    return source.slice(0, insertPos) + childMarkup + source.slice(insertPos);
  }
  const tagName = getTagName(opening.name);
  const reopened = source.slice(opening.start, opening.end).replace(/\s*\/>$/, ">");
  return source.slice(0, opening.start) + `${reopened}${childMarkup}</${tagName}>` + source.slice(opening.end);
}
function injectRootMetaIntoSource(source, filename, root) {
  let parseResult;
  try {
    parseResult = (0, import_oxc_parser3.parseSync)(filename, source);
  } catch {
    return source;
  }
  const metaTag = `<meta name="${ROOT_META_NAME}" content={${JSON.stringify(root)}} />`;
  const headElement = findFirstJSXElement(parseResult.program, "head");
  if (headElement) {
    if (hasRootMetaChild(headElement)) return source;
    return injectFirstChildMarkup(source, headElement, metaTag);
  }
  const htmlElement = findFirstJSXElement(parseResult.program, "html");
  if (htmlElement) {
    return injectFirstChildMarkup(
      source,
      htmlElement,
      `<head>${metaTag}</head>`
    );
  }
  return source;
}
function injectRootMetaIntoAnnotatedSource(annotatedSource, filename, root) {
  return injectRootMetaIntoSource(annotatedSource, filename, root);
}
function loadUserComputeProps(modulePath, canonicalRoot) {
  const resolved = (0, import_node_path2.resolve)(canonicalRoot, modulePath);
  const requireFromRoot = (0, import_node_module.createRequire)(`${canonicalRoot}/__tempo_loader__.js`);
  const mod = requireFromRoot(resolved);
  const fn = mod && (mod.default ?? mod);
  if (typeof fn !== "function") {
    return null;
  }
  return fn;
}
function readOptions() {
  if (typeof this.getOptions === "function") {
    return this.getOptions() ?? {};
  }
  if (typeof this.query === "string") {
    return {};
  }
  return this.query ?? {};
}
function nextjsLoader(source) {
  this.cacheable?.(true);
  if (!process.env.TEMPO) return source;
  const cleanPath = stripQueryAndHash(this.resourcePath ?? "");
  if (!/\.(tsx|jsx)$/.test(cleanPath)) return source;
  if (isInNodeModules(cleanPath)) return source;
  const options = readOptions.call(this);
  try {
    if (cachedCanonicalRoot === null) {
      cachedCanonicalRoot = resolveCanonicalRoot(this.rootContext);
    }
    const canonicalRoot = cachedCanonicalRoot;
    const userModule = options.computeAnnotatedPropsModule;
    if (userModule && userModule !== cachedComputeModulePath) {
      try {
        cachedUserComputeProps = loadUserComputeProps(
          userModule,
          canonicalRoot
        );
      } catch {
        cachedUserComputeProps = null;
      }
      cachedComputeModulePath = userModule;
    } else if (cachedUserComputeProps === void 0) {
      cachedUserComputeProps = null;
    }
    const userComputeProps = cachedUserComputeProps;
    const filepath = (0, import_node_path2.relative)(canonicalRoot, cleanPath);
    const modifiedtsMs = Math.trunc((0, import_node_fs2.statSync)(cleanPath).mtimeMs);
    const sourceVersion = sourceVersionHash(source);
    const annotated = annotateFile(
      source,
      cleanPath,
      (info) => computeTempoAnnotatedProps({
        info,
        filepath,
        modifiedtsMs,
        sourceVersion,
        computeAnnotatedProps: userComputeProps ?? void 0
      })
    );
    let finalSource = annotated.annotatedSource;
    if (options.autoInjectRootMeta !== false) {
      finalSource = injectRootMetaIntoAnnotatedSource(
        finalSource,
        cleanPath,
        canonicalRoot
      );
    }
    if (finalSource === source) return source;
    return finalSource;
  } catch {
    return source;
  }
}
var nextjs_loader_default = nextjsLoader;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  _resetLoaderState,
  injectRootMetaIntoAnnotatedSource,
  injectRootMetaIntoSource,
  nextjsLoader
});
