// tempo-sdk/dist/chunk-6Q4I3RFN.js
// canvas/parser.ts
import { parseSync } from "oxc-parser";
var SDK_IMPORT_SOURCES = /* @__PURE__ */ new Set(["tempo-sdk/canvas"]);
var SDK_COMPONENTS = /* @__PURE__ */ new Set([
  "Canvas",
  "Storyboard",
  "LegacyStoryboardV1",
  "RouteStoryboard",
  "HtmlStoryboard"
]);
var DEFAULT_LAYOUT = {
  x: 0,
  y: 0,
  width: 600,
  height: 400
};
function parseCanvasFile(filePath, source) {
  const errors = [];
  let program;
  try {
    const result = parseSync(filePath, source);
    program = result.program;
    for (const error of result.errors ?? []) {
      errors.push({
        message: error.message ?? "Unable to parse canvas source",
        span: toSpan(error) ?? toSpan(error?.labels?.[0])
      });
    }
  } catch (error) {
    return {
      filePath,
      pageName: fallbackPageName(filePath),
      imports: [],
      storyboards: [],
      errors: [
        {
          message: error instanceof Error ? error.message : "Unable to parse canvas source"
        }
      ]
    };
  }
  const imports = readImports(program);
  const { legacyObjects, stringBindings: objectStringBindings } = readObjectExpressionBindings(program, imports, source);
  const storyboards = [];
  let canvasElement = null;
  walk(program, (node) => {
    if (node.type !== "JSXElement") return;
    const component = resolveSdkComponent(node.openingElement?.name, imports);
    if (component === "Canvas" && !canvasElement) {
      canvasElement = node;
      return;
    }
    if (component === "Storyboard" || component === "LegacyStoryboardV1" || component === "RouteStoryboard" || component === "HtmlStoryboard") {
      storyboards.push(
        readStoryboard(node, component, imports, legacyObjects, source, errors)
      );
    }
  });
  if (!canvasElement) {
    errors.push({ message: "No <Canvas> declaration found in canvas file" });
  } else {
    errors.push(...validateCanvasChildren(canvasElement, imports));
  }
  const openingElement = canvasElement?.openingElement;
  const name = readResolvableStringAttribute(
    openingElement,
    "name",
    objectStringBindings
  );
  const path = readStringAttribute(openingElement, "path");
  const backgroundColor = readResolvableStringAttribute(
    openingElement,
    "backgroundColor",
    objectStringBindings
  ) ?? readResolvableStringAttribute(
    openingElement,
    "background",
    objectStringBindings
  );
  if (openingElement && !name) {
    errors.push({
      message: '<Canvas> must include a literal "name" prop',
      span: toSpan(openingElement)
    });
  }
  return {
    filePath,
    canvasPath: path?.value,
    pageName: name?.value ?? fallbackPageName(filePath),
    backgroundColor: backgroundColor?.value,
    canvasElementSpan: toSpan(canvasElement),
    canvasOpeningElementSpan: toSpan(openingElement),
    canvasChildrenInsertionOffset: typeof canvasElement?.closingElement?.start === "number" ? canvasElement.closingElement.start : void 0,
    nameSpan: name?.valueSpan,
    nameAttributeSpan: name?.attributeSpan,
    pathSpan: path?.valueSpan,
    pathAttributeSpan: path?.attributeSpan,
    backgroundColorSpan: backgroundColor?.valueSpan,
    backgroundColorAttributeSpan: backgroundColor?.attributeSpan,
    imports: imports.all,
    storyboards,
    errors
  };
}
function readImports(program) {
  const all = [];
  const byLocalName = /* @__PURE__ */ new Map();
  const sdkLocalNames = /* @__PURE__ */ new Map();
  const sdkNamespaces = /* @__PURE__ */ new Set();
  for (const statement of program?.body ?? []) {
    if (statement?.type !== "ImportDeclaration" || typeof statement.source?.value !== "string") {
      continue;
    }
    const source = statement.source.value;
    const isSdkImport = isSdkImportSource(source);
    for (const specifier of statement.specifiers ?? []) {
      const localName = specifier.local?.name;
      if (typeof localName !== "string") continue;
      const importedName = readImportedName(specifier);
      const imported = {
        localName,
        source,
        importedName,
        kind: readImportKind(specifier),
        importKind: specifier.importKind === "type" ? "type" : "value",
        span: toSpan(specifier) ?? toSpan(statement) ?? { start: 0, end: 0 },
        declarationSpan: toSpan(statement) ?? { start: 0, end: 0 },
        specifierCount: statement.specifiers?.length ?? 0,
        namedSpecifierCount: (statement.specifiers ?? []).filter(
          (item) => item?.type === "ImportSpecifier"
        ).length,
        hasDefaultSpecifier: (statement.specifiers ?? []).some(
          (item) => item?.type === "ImportDefaultSpecifier"
        ),
        hasNamespaceSpecifier: (statement.specifiers ?? []).some(
          (item) => item?.type === "ImportNamespaceSpecifier"
        )
      };
      all.push(imported);
      byLocalName.set(localName, imported);
      if (!isSdkImport) continue;
      if (specifier.type === "ImportNamespaceSpecifier") {
        sdkNamespaces.add(localName);
        continue;
      }
      if (SDK_COMPONENTS.has(importedName)) {
        sdkLocalNames.set(localName, importedName);
      }
    }
  }
  return { all, byLocalName, sdkLocalNames, sdkNamespaces };
}
function readObjectExpressionBindings(program, imports, source) {
  const legacyObjects = /* @__PURE__ */ new Map();
  const stringBindings = /* @__PURE__ */ new Map();
  walk(program, (node) => {
    if (node?.type !== "VariableDeclarator" || node.id?.type !== "Identifier") {
      return;
    }
    const objectExpression = unwrapTsExpression(node.init);
    if (objectExpression?.type !== "ObjectExpression") return;
    const bindingName = node.id.name;
    legacyObjects.set(
      bindingName,
      readLegacyStoryboardObject(bindingName, objectExpression, imports, source)
    );
    const stringProperties = readObjectStringProperties(objectExpression);
    if (stringProperties.size > 0) {
      stringBindings.set(bindingName, stringProperties);
    }
  });
  return { legacyObjects, stringBindings };
}
function readLegacyStoryboardObject(bindingName, objectExpression, imports, source) {
  const errors = [];
  const nameProperty = getObjectProperty(objectExpression, "name");
  const name = nameProperty ? readStringLiteralValue(nameProperty.value) : null;
  const layoutProperty = getObjectProperty(objectExpression, "layout");
  const layout = layoutProperty ? readLayoutExpression(layoutProperty.value, "LegacyStoryboardV1") : {
    value: DEFAULT_LAYOUT,
    errors: [
      {
        message: '<LegacyStoryboardV1> storyboard object must include a literal "layout" object',
        span: toSpan(objectExpression)
      }
    ]
  };
  const renderProperty = getObjectProperty(objectExpression, "render");
  const component = readRenderComponentReference(
    renderProperty?.value,
    imports,
    source
  );
  if (nameProperty && !name) {
    errors.push({
      message: "<LegacyStoryboardV1> storyboard.name must be a string literal when set",
      span: toSpan(nameProperty)
    });
  }
  errors.push(...layout.errors);
  return {
    bindingName,
    displayName: name?.value,
    component: component ?? void 0,
    objectSpan: toSpan(objectExpression),
    objectClose: readLayoutClose(objectExpression),
    nameSpan: name?.span,
    namePropertySpan: toSpan(nameProperty),
    layoutPropertySpan: toSpan(layoutProperty),
    layout: layout.value,
    layoutSpans: layout.spans,
    errors
  };
}
function readObjectStringProperties(objectExpression) {
  const properties = /* @__PURE__ */ new Map();
  for (const property of objectExpression.properties ?? []) {
    if (property?.type !== "Property" || property.computed) continue;
    const propertyName = readPropertyKeyName(property.key);
    const value = readStringLiteralValue(property.value);
    if (!propertyName || !value) continue;
    properties.set(propertyName, value);
  }
  return properties;
}
function readStoryboard(element, component, imports, legacyObjects, source, globalErrors) {
  const openingElement = element.openingElement;
  const errors = [];
  const id = readStringAttribute(openingElement, "id");
  const name = readStringAttribute(openingElement, "name");
  const legacyReference = component === "LegacyStoryboardV1" ? readExpressionReference(openingElement, "storyboard", source) : void 0;
  const legacyObject = legacyReference?.expression && isIdentifierExpression(legacyReference.expression) ? legacyObjects.get(legacyReference.expression) : void 0;
  const layout = component === "LegacyStoryboardV1" ? {
    value: legacyObject?.layout ?? DEFAULT_LAYOUT,
    spans: legacyObject?.layoutSpans,
    attributeSpan: void 0,
    errors: legacyObject?.errors ?? (legacyReference ? [
      {
        message: "<LegacyStoryboardV1> storyboard prop must reference a same-file object literal",
        span: legacyReference.span
      }
    ] : [
      {
        message: '<LegacyStoryboardV1> must include a "storyboard" expression prop',
        span: toSpan(openingElement)
      }
    ])
  } : readLayout(openingElement, component);
  const kind = component === "Storyboard" || component === "LegacyStoryboardV1" ? "component" : component === "RouteStoryboard" ? "route" : "html";
  if (!id) {
    errors.push({
      message: `<${component}> must include a literal "id" prop`,
      span: toSpan(openingElement)
    });
  }
  if (layout.errors.length > 0) {
    errors.push(...layout.errors);
  }
  const storyboard = {
    storyboardKey: id?.value ?? "",
    displayName: name?.value ?? legacyObject?.displayName,
    kind,
    layout: layout.value,
    editable: Boolean(id) && layout.errors.length === 0,
    jsxElementSpan: toSpan(element) ?? { start: 0, end: 0 },
    openingElementSpan: toSpan(openingElement),
    idSpan: id?.valueSpan,
    idAttributeSpan: id?.attributeSpan,
    nameSpan: name?.valueSpan ?? legacyObject?.nameSpan,
    nameAttributeSpan: name?.attributeSpan ?? legacyObject?.namePropertySpan,
    layoutAttributeSpan: layout.attributeSpan,
    layoutSpans: layout.spans,
    props: readExpressionReference(openingElement, "props", source),
    component: legacyObject?.component,
    legacyV1: legacyReference ? {
      storyboard: legacyReference,
      objectName: legacyObject?.bindingName,
      objectSpan: legacyObject?.objectSpan,
      objectClose: legacyObject?.objectClose,
      namePropertySpan: legacyObject?.namePropertySpan,
      layoutPropertySpan: legacyObject?.layoutPropertySpan
    } : void 0,
    errors
  };
  if (component === "Storyboard") {
    const componentReference = readComponentReference(
      openingElement,
      imports,
      source
    );
    if (componentReference) {
      storyboard.component = componentReference;
    } else {
      errors.push({
        message: '<Storyboard> must include a "component" expression prop',
        span: toSpan(openingElement)
      });
    }
  }
  if (component === "RouteStoryboard") {
    const route = readStringAttribute(openingElement, "route");
    if (route) {
      storyboard.route = route.value;
      storyboard.routeSpan = route.valueSpan;
      storyboard.routeAttributeSpan = route.attributeSpan;
    } else {
      errors.push({
        message: '<RouteStoryboard> must include a literal "route" prop',
        span: toSpan(openingElement)
      });
    }
  }
  if (component === "HtmlStoryboard") {
    const filepath = readStringAttribute(openingElement, "filepath");
    const src = readStringAttribute(openingElement, "src");
    storyboard.filepath = filepath?.value;
    storyboard.filepathSpan = filepath?.valueSpan;
    storyboard.filepathAttributeSpan = filepath?.attributeSpan;
    storyboard.src = src?.value;
    storyboard.srcSpan = src?.valueSpan;
    storyboard.srcAttributeSpan = src?.attributeSpan;
    if (!filepath && !src) {
      errors.push({
        message: '<HtmlStoryboard> must include a literal "filepath" or "src" prop',
        span: toSpan(openingElement)
      });
    }
  }
  globalErrors.push(...errors);
  return storyboard;
}
function validateCanvasChildren(canvasElement, imports) {
  const errors = [];
  for (const child of canvasElement.children ?? []) {
    if (!child) continue;
    if (child.type === "JSXText") {
      if (typeof child.value === "string" && child.value.trim() === "") {
        continue;
      }
    }
    if (child.type === "JSXExpressionContainer") {
      if (child.expression?.type === "JSXEmptyExpression") {
        continue;
      }
      errors.push({
        message: "<Canvas> children must be static storyboard declarations, not expressions",
        span: toSpan(child)
      });
      continue;
    }
    if (child.type === "JSXElement") {
      const component = resolveSdkComponent(
        child.openingElement?.name,
        imports
      );
      if (component === "Storyboard" || component === "LegacyStoryboardV1" || component === "RouteStoryboard" || component === "HtmlStoryboard") {
        continue;
      }
    }
    errors.push({
      message: "<Canvas> children must be <Storyboard>, <RouteStoryboard>, <HtmlStoryboard>, or <LegacyStoryboardV1> declarations",
      span: toSpan(child)
    });
  }
  return errors;
}
function readComponentReference(openingElement, imports, source) {
  const attribute = getAttribute(openingElement, "component");
  const expression = unwrapAttributeExpression(attribute?.value);
  const span = toSpan(expression);
  if (!expression || !span) return null;
  const localName = expression.type === "Identifier" ? expression.name : expression.type === "MemberExpression" && expression.object?.type === "Identifier" ? expression.object.name : void 0;
  return {
    expression: source.slice(span.start, span.end),
    span,
    attributeSpan: toSpan(attribute),
    localName,
    import: localName ? imports.byLocalName.get(localName) : void 0
  };
}
function readRenderComponentReference(renderValue, imports, source) {
  if (renderValue?.type !== "ArrowFunctionExpression") return null;
  const body = unwrapTsExpression(renderValue.body);
  if (body?.type !== "JSXElement") return null;
  const name = body.openingElement?.name;
  if (name?.type !== "JSXIdentifier") return null;
  const span = toSpan(name);
  if (!span || typeof name.name !== "string") return null;
  return {
    expression: source.slice(span.start, span.end),
    span,
    localName: name.name,
    import: imports.byLocalName.get(name.name)
  };
}
function readExpressionReference(openingElement, name, source) {
  const attribute = getAttribute(openingElement, name);
  const expression = unwrapAttributeExpression(attribute?.value);
  const span = toSpan(expression);
  if (!expression || !span) return void 0;
  return {
    expression: source.slice(span.start, span.end),
    span,
    attributeSpan: toSpan(attribute)
  };
}
function isIdentifierExpression(expression) {
  return /^[A-Za-z_$][\w$]*$/.test(expression);
}
function readLayout(openingElement, componentName) {
  const errors = [];
  const attribute = getAttribute(openingElement, "layout");
  const expression = unwrapAttributeExpression(attribute?.value);
  if (!expression) {
    return {
      value: DEFAULT_LAYOUT,
      attributeSpan: toSpan(attribute),
      errors: [
        {
          message: `<${componentName}> must include a literal "layout" object prop`,
          span: toSpan(openingElement)
        }
      ]
    };
  }
  const layout = readLayoutExpression(expression, componentName);
  return {
    ...layout,
    attributeSpan: toSpan(attribute)
  };
}
function readLayoutExpression(expression, componentName) {
  const errors = [];
  if (expression.type !== "ObjectExpression") {
    return {
      value: DEFAULT_LAYOUT,
      spans: { layoutClose: readLayoutClose(expression) },
      errors: [
        {
          message: `<${componentName}> layout must be an object literal`,
          span: toSpan(expression)
        }
      ]
    };
  }
  const spans = {
    layoutClose: readLayoutClose(expression)
  };
  const layout = { ...DEFAULT_LAYOUT };
  for (const key of ["x", "y", "width", "height"]) {
    const property = getObjectProperty(expression, key);
    const value = property ? readNumberLiteral(property.value) : null;
    if (value) {
      layout[key] = value.value;
      spans[key] = value.span;
    } else {
      errors.push({
        message: `<${componentName}> layout.${key} must be a number literal`,
        span: toSpan(property ?? expression)
      });
    }
  }
  const zIndex = getObjectProperty(expression, "zIndex");
  if (zIndex) {
    const value = readNumberLiteral(zIndex.value);
    if (value) {
      layout.zIndex = value.value;
      spans.zIndex = value.span;
    } else {
      errors.push({
        message: `<${componentName}> layout.zIndex must be a number literal`,
        span: toSpan(zIndex)
      });
    }
  }
  const intrinsicSizing = getObjectProperty(expression, "intrinsicSizing");
  if (intrinsicSizing) {
    if (intrinsicSizing.value?.type === "Literal" && intrinsicSizing.value.value === "root-element") {
      layout.intrinsicSizing = "root-element";
      spans.intrinsicSizing = toSpan(intrinsicSizing.value);
    } else {
      errors.push({
        message: `<${componentName}> layout.intrinsicSizing must be "root-element" when set`,
        span: toSpan(intrinsicSizing)
      });
    }
  }
  return { value: layout, spans, errors };
}
function readStringLiteralValue(node) {
  return node?.type === "Literal" && typeof node.value === "string" ? { value: node.value, span: toSpan(node) ?? { start: 0, end: 0 } } : null;
}
function unwrapTsExpression(node) {
  let current = node;
  while (current?.type === "TSSatisfiesExpression" || current?.type === "TSAsExpression" || current?.type === "TSNonNullExpression" || current?.type === "ParenthesizedExpression") {
    current = current.expression;
  }
  return current;
}
function readResolvableStringAttribute(openingElement, name, bindings) {
  const literal = readStringAttribute(openingElement, name);
  if (literal) return literal;
  const attribute = getAttribute(openingElement, name);
  const expression = unwrapAttributeExpression(attribute?.value);
  const expressionSpan = toSpan(expression);
  const attributeSpan = toSpan(attribute);
  if (!expression || !expressionSpan || !attributeSpan) return null;
  const resolved = resolveStringMemberExpression(expression, bindings);
  return resolved ? {
    value: resolved.value,
    valueSpan: expressionSpan,
    attributeSpan
  } : null;
}
function resolveStringMemberExpression(expression, bindings) {
  if (expression?.type !== "MemberExpression") return null;
  if (expression.object?.type !== "Identifier") return null;
  const propertyName = readPropertyKeyName(expression.property);
  if (!propertyName) return null;
  return bindings.get(expression.object.name)?.get(propertyName) ?? null;
}
function readStringAttribute(openingElement, name) {
  const attribute = getAttribute(openingElement, name);
  if (!attribute) return null;
  const value = attribute.value;
  if (value?.type === "Literal" && typeof value.value === "string") {
    return {
      value: value.value,
      valueSpan: toSpan(value) ?? toSpan(attribute) ?? { start: 0, end: 0 },
      attributeSpan: toSpan(attribute) ?? { start: 0, end: 0 }
    };
  }
  const expression = unwrapAttributeExpression(value);
  if (expression?.type === "Literal" && typeof expression.value === "string") {
    return {
      value: expression.value,
      valueSpan: toSpan(expression) ?? toSpan(attribute) ?? { start: 0, end: 0 },
      attributeSpan: toSpan(attribute) ?? { start: 0, end: 0 }
    };
  }
  return null;
}
function getAttribute(openingElement, name) {
  for (const attribute of openingElement?.attributes ?? []) {
    if (attribute?.type === "JSXAttribute" && attribute.name?.type === "JSXIdentifier" && attribute.name.name === name) {
      return attribute;
    }
  }
  return null;
}
function unwrapAttributeExpression(value) {
  if (!value) return null;
  if (value.type === "JSXExpressionContainer") {
    return value.expression;
  }
  return value;
}
function readPropertyKeyName(key) {
  if (key?.type === "Identifier") return key.name;
  if (key?.type === "Literal" && typeof key.value === "string") {
    return key.value;
  }
  return null;
}
function getObjectProperty(objectExpression, name) {
  if (objectExpression?.type !== "ObjectExpression") return null;
  for (const property of objectExpression.properties ?? []) {
    if (property?.type !== "Property" || property.computed) continue;
    const propertyName = readPropertyKeyName(property.key);
    if (propertyName === name) return property;
  }
  return null;
}
function readNumberLiteral(node) {
  if (node?.type === "Literal" && typeof node.value === "number" && Number.isFinite(node.value)) {
    return { value: node.value, span: toSpan(node) ?? { start: 0, end: 0 } };
  }
  if (node?.type === "UnaryExpression" && node.operator === "-" && node.argument?.type === "Literal" && typeof node.argument.value === "number" && Number.isFinite(node.argument.value)) {
    return {
      value: -node.argument.value,
      span: toSpan(node) ?? { start: 0, end: 0 }
    };
  }
  return null;
}
function resolveSdkComponent(nameNode, imports) {
  if (nameNode?.type === "JSXIdentifier") {
    return imports.sdkLocalNames.get(nameNode.name) ?? null;
  }
  if (nameNode?.type === "JSXMemberExpression" && nameNode.object?.type === "JSXIdentifier" && nameNode.property?.type === "JSXIdentifier" && imports.sdkNamespaces.has(nameNode.object.name) && SDK_COMPONENTS.has(nameNode.property.name)) {
    return nameNode.property.name;
  }
  return null;
}
function readImportedName(specifier) {
  if (specifier.type === "ImportDefaultSpecifier") return "default";
  if (specifier.type === "ImportNamespaceSpecifier") return "*";
  return specifier.imported?.name ?? specifier.local?.name ?? "unknown";
}
function readImportKind(specifier) {
  if (specifier.type === "ImportDefaultSpecifier") return "default";
  if (specifier.type === "ImportNamespaceSpecifier") return "namespace";
  return "named";
}
function isSdkImportSource(source) {
  return SDK_IMPORT_SOURCES.has(source) || source.endsWith("/tempo-sdk/canvas");
}
function readLayoutClose(node) {
  return typeof node?.end === "number" && Number.isFinite(node.end) ? node.end - 1 : void 0;
}
function fallbackPageName(filePath) {
  const normalized = filePath.replace(/\\/g, "/");
  const segments = normalized.split("/").filter(Boolean);
  const fileName = segments.length > 0 ? segments[segments.length - 1] : filePath;
  const parentName = segments.length > 1 ? segments[segments.length - 2] : fileName;
  const baseName = /^index\.canvas\.[tj]sx?$/.test(fileName) ? parentName : fileName.replace(/\.canvas\.[tj]sx?$/, "");
  return baseName.split(/[-_\s]+/).filter(Boolean).map((part) => part.charAt(0).toUpperCase() + part.slice(1)).join(" ");
}
function toSpan(node) {
  return typeof node?.start === "number" && Number.isFinite(node.start) && typeof node?.end === "number" && Number.isFinite(node.end) ? { start: node.start, end: node.end } : void 0;
}
function walk(node, visitor, seen = /* @__PURE__ */ new Set()) {
  if (!node || typeof node !== "object" || seen.has(node)) return;
  seen.add(node);
  visitor(node);
  for (const value of Object.values(node)) {
    if (!value) continue;
    if (Array.isArray(value)) {
      for (const item of value) {
        walk(item, visitor, seen);
      }
    } else if (typeof value === "object") {
      walk(value, visitor, seen);
    }
  }
}

export {
  parseCanvasFile,
  unwrapTsExpression,
  fallbackPageName
};
