// tempo-sdk/dist/expo-babel-plugin.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// expo-babel-plugin.ts
var expo_babel_plugin_exports = {};
__export(expo_babel_plugin_exports, {
  default: () => tempoExpoBabelPlugin,
  tempoExpoBabelPlugin: () => tempoExpoBabelPlugin
});
module.exports = __toCommonJS(expo_babel_plugin_exports);
var import_node_fs2 = require("fs");
var import_node_path2 = require("path");
var import_pluginutils = require("@rollup/pluginutils");

// annotation-strategy.ts
var import_node_fs = require("fs");
var import_node_path = require("path");

// ../../modules/annotation/source-version.ts
function sourceVersionHash(content) {
  let djb2 = 5381;
  let sdbm = 0;
  for (let i = 0; i < content.length; i++) {
    const c = content.charCodeAt(i);
    djb2 = (djb2 << 5) + djb2 ^ c;
    sdbm = c + (sdbm << 6) + (sdbm << 16) - sdbm | 0;
  }
  return toHex8(djb2) + toHex8(sdbm);
}
function toHex8(n) {
  return (n >>> 0).toString(16).padStart(8, "0");
}

// annotation-strategy.ts
var DEFAULT_EXCLUDE = ["**/node_modules/**"];
var HTML_VOID_ELEMENTS = /* @__PURE__ */ new Set([
  "area",
  "base",
  "br",
  "col",
  "embed",
  "hr",
  "img",
  "input",
  "link",
  "meta",
  "param",
  "source",
  "track",
  "wbr"
]);
function cleanModuleId(id) {
  return id.replace(/[?#].*$/, "");
}
function resolveCanonicalRoot(root) {
  try {
    return import_node_fs.realpathSync.native(root);
  } catch {
    return root;
  }
}
function isIntrinsicElement(tagName) {
  return tagName.length > 0 && tagName[0] === tagName[0].toLowerCase();
}
function hasJsxAttribute(info, attrName) {
  for (const attr of info.element.openingElement.attributes) {
    const name = attr.name;
    if (name?.name === attrName) return true;
  }
  return false;
}
function defaultComputeAnnotatedProps(info, filepath, modifiedtsMs, sourceVersion) {
  return {
    "data-tempo-filepath": filepath,
    "data-tempo-position": info.element.openingElement.start,
    "data-tempo-modifiedts": modifiedtsMs,
    // Content-hash identity of the file this element was annotated from. The
    // canvas reconcile loop equality-compares it against the write pipeline's
    // computed version to decide whether a capture reflects a given edit.
    "data-tempo-source-version": sourceVersion
  };
}
function computeTempoAnnotatedProps(params) {
  const {
    info,
    filepath,
    modifiedtsMs,
    sourceVersion,
    computeAnnotatedProps,
    elementSupportOverrides
  } = params;
  const userProps = computeAnnotatedProps ? computeAnnotatedProps(info, filepath, modifiedtsMs, sourceVersion) : defaultComputeAnnotatedProps(info, filepath, modifiedtsMs, sourceVersion);
  if (isIntrinsicElement(info.tagName)) {
    return {
      ...userProps,
      "data-tempo-supports-children": hasJsxAttribute(
        info,
        "data-tempo-supports-children"
      ) ? void 0 : !HTML_VOID_ELEMENTS.has(info.tagName)
    };
  }
  const override = elementSupportOverrides?.get(info.tagName);
  if (override) {
    return {
      ...userProps,
      "data-tempo-supports-children": hasJsxAttribute(
        info,
        "data-tempo-supports-children"
      ) ? void 0 : override.supportsChildren
    };
  }
  return userProps;
}
function getFileAnnotationContext(params) {
  const rawCleanFilename = cleanModuleId(params.filename);
  let cleanFilename = rawCleanFilename;
  if (params.canonicalizeFilename) {
    try {
      cleanFilename = import_node_fs.realpathSync.native(rawCleanFilename);
    } catch {
      cleanFilename = rawCleanFilename;
    }
  }
  return {
    cleanFilename,
    filepath: (0, import_node_path.relative)(params.root, cleanFilename),
    modifiedtsMs: Math.trunc((0, import_node_fs.statSync)(cleanFilename).mtimeMs),
    // Hashed once per file per transform (the context is built once, reused
    // for every element) so the per-element funnel just forwards the token.
    sourceVersion: sourceVersionHash(params.source)
  };
}

// expo-babel-plugin.ts
var EXPO_DEFAULT_INCLUDE = ["**/*.tsx", "**/*.jsx", "**/*.js"];
var FRAGMENT_TAG_NAMES = /* @__PURE__ */ new Set(["Fragment", "React.Fragment"]);
var REACT_NATIVE_HOST_COMPONENTS = /* @__PURE__ */ new Set([
  "ActivityIndicator",
  "Button",
  "FlatList",
  "Image",
  "ImageBackground",
  "KeyboardAvoidingView",
  "Modal",
  "Pressable",
  "RefreshControl",
  "SafeAreaView",
  "ScrollView",
  "SectionList",
  "Switch",
  "Text",
  "TextInput",
  "TouchableHighlight",
  "TouchableNativeFeedback",
  "TouchableOpacity",
  "TouchableWithoutFeedback",
  "View",
  "VirtualizedList"
]);
var REACT_NATIVE_NO_CHILDREN_COMPONENTS = /* @__PURE__ */ new Set([
  "ActivityIndicator",
  "Image",
  "RefreshControl",
  "Switch"
]);
var REACT_NATIVE_SUPPORT_OVERRIDES = new Map(
  [...REACT_NATIVE_HOST_COMPONENTS].map((tagName) => [
    tagName,
    { supportsChildren: !REACT_NATIVE_NO_CHILDREN_COMPONENTS.has(tagName) }
  ])
);
function hasParserPlugin(plugins, name) {
  return plugins.some(
    (plugin) => Array.isArray(plugin) ? plugin[0] === name : plugin === name
  );
}
function getBabelTagName(name) {
  switch (name.type) {
    case "JSXIdentifier":
      return name.name;
    case "JSXNamespacedName":
      return `${name.namespace.name}:${name.name.name}`;
    case "JSXMemberExpression":
      return `${getBabelTagName(name.object)}.${name.property.name}`;
  }
}
function hasAttribute(node, attrName) {
  return node.attributes.some(
    (attr) => attr.type === "JSXAttribute" && attr.name.name === attrName
  );
}
function propValueToExpression(api, value) {
  if (typeof value === "function") {
    return api.template.expression(value.toString())();
  }
  return api.types.valueToNode(value);
}
function dataAttrToDataSetKey(attrName) {
  return attrName.replace(/^data-/, "").replace(/-([a-z])/g, (_, char) => char.toUpperCase());
}
function findDataSetAttribute(openingElement) {
  return openingElement.attributes.find(
    (attr) => attr.type === "JSXAttribute" && attr.name.name === "dataSet"
  );
}
function addAnnotations(api, tagName, openingElement, props) {
  const t = api.types;
  if (isIntrinsicElement(tagName)) {
    for (const [key, value] of Object.entries(props)) {
      if (value === void 0 || hasAttribute(openingElement, key)) continue;
      openingElement.attributes.push(
        t.jsxAttribute(
          t.jsxIdentifier(key),
          t.jsxExpressionContainer(propValueToExpression(api, value))
        )
      );
    }
    return;
  }
  const dataSetProps = [];
  for (const [key, value] of Object.entries(props)) {
    if (value === void 0) continue;
    dataSetProps.push(
      t.objectProperty(
        t.identifier(dataAttrToDataSetKey(key)),
        propValueToExpression(api, value)
      )
    );
  }
  if (dataSetProps.length === 0) return;
  const existing = findDataSetAttribute(openingElement);
  if (!existing) {
    openingElement.attributes.push(
      t.jsxAttribute(
        t.jsxIdentifier("dataSet"),
        t.jsxExpressionContainer(t.objectExpression(dataSetProps))
      )
    );
    return;
  }
  if (existing.value?.type === "JSXExpressionContainer" && existing.value.expression.type !== "JSXEmptyExpression") {
    existing.value.expression = t.objectExpression([
      t.spreadElement(existing.value.expression),
      ...dataSetProps
    ]);
  }
}
var ROOT_META_NAME = "tempo-project-root";
function injectProjectRootMeta(api, programPath, root) {
  const setter = api.template.statement(`
    if (typeof document !== "undefined") {
      var __tempoRootMeta = document.querySelector('meta[name="${ROOT_META_NAME}"]');
      if (!__tempoRootMeta) {
        __tempoRootMeta = document.createElement("meta");
        __tempoRootMeta.setAttribute("name", "${ROOT_META_NAME}");
        document.head.appendChild(__tempoRootMeta);
      }
      __tempoRootMeta.setAttribute("content", ${JSON.stringify(root)});
    }
  `)();
  programPath.pushContainer("body", setter);
}
function resolveBabelRoot(state, options, filename) {
  return resolveCanonicalRoot(
    (0, import_node_path2.resolve)(
      options.root ?? findNearestPackageRoot(filename) ?? state.cwd ?? state.file.opts.cwd ?? state.file.opts.root ?? process.cwd()
    )
  );
}
function findNearestPackageRoot(filename) {
  let current = (0, import_node_path2.dirname)(filename);
  while (true) {
    if ((0, import_node_fs2.existsSync)((0, import_node_path2.resolve)(current, "package.json"))) return current;
    const parent = (0, import_node_path2.dirname)(current);
    if (parent === current) return null;
    current = parent;
  }
}
function resolveFilename(state) {
  const filename = state.filename ?? state.file.opts.filename;
  if (!filename) return null;
  return (0, import_node_path2.resolve)(cleanModuleId(filename));
}
function tempoExpoBabelPlugin(api, rawOptions = {}) {
  api.assertVersion(7);
  const enabled = rawOptions.enabled ?? !!process.env.TEMPO;
  const include = rawOptions.include === void 0 ? EXPO_DEFAULT_INCLUDE : rawOptions.include;
  const exclude = rawOptions.exclude === void 0 ? DEFAULT_EXCLUDE : rawOptions.exclude;
  const filter = (0, import_pluginutils.createFilter)(include, exclude);
  return {
    name: "tempo-expo-annotate",
    manipulateOptions(opts, parserOpts) {
      parserOpts.plugins ?? (parserOpts.plugins = []);
      if (!hasParserPlugin(parserOpts.plugins, "jsx")) {
        parserOpts.plugins.push("jsx");
      }
      const filename = opts.filename ?? "";
      if (/\.(ts|tsx)$/.test(filename) && !hasParserPlugin(parserOpts.plugins, "typescript")) {
        parserOpts.plugins.push("typescript");
      }
    },
    visitor: {
      Program: {
        enter(path, state) {
          if (!enabled) return;
          const options = { ...rawOptions, ...state.opts };
          const filename = resolveFilename(state);
          if (!filename || !filter(filename)) return;
          const source = state.file.code;
          const root = resolveBabelRoot(state, options, filename);
          let context;
          try {
            context = getFileAnnotationContext({
              source,
              filename,
              root,
              canonicalizeFilename: true
            });
          } catch {
            return;
          }
          if (context.filepath === "index.js") {
            injectProjectRootMeta(api, path, root);
          }
          path.traverse({
            JSXElement(elementPath) {
              const element = elementPath.node;
              const tagName = getBabelTagName(element.openingElement.name);
              if (FRAGMENT_TAG_NAMES.has(tagName)) return;
              const info = {
                element,
                tagName
              };
              const props = computeTempoAnnotatedProps({
                info,
                filepath: context.filepath,
                modifiedtsMs: context.modifiedtsMs,
                sourceVersion: context.sourceVersion,
                computeAnnotatedProps: options.computeAnnotatedProps,
                elementSupportOverrides: REACT_NATIVE_SUPPORT_OVERRIDES
              });
              addAnnotations(api, tagName, element.openingElement, props);
            }
          });
        }
      }
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  tempoExpoBabelPlugin
});
