// tempo-sdk/dist/index.js
// index.ts
import { realpathSync as realpathSync2, statSync as statSync2 } from "fs";
import { relative as relative2 } from "path";

// ../../modules/annotation/annotate-file.ts
import { parseSync as parseSync2 } from "oxc-parser";

// ../../modules/file-parser/utils.ts
function getTagName(name) {
  switch (name.type) {
    case "JSXIdentifier":
      return name.name;
    case "JSXNamespacedName":
      return `${name.namespace.name}:${name.name.name}`;
    case "JSXMemberExpression":
      return getJSXMemberExpressionName(name);
  }
}
function getJSXMemberExpressionName(node) {
  const object = node.object.type === "JSXIdentifier" ? node.object.name : getJSXMemberExpressionName(node.object);
  return `${object}.${node.property.name}`;
}

// ../../modules/file-parser/walk.ts
function walk(node, callbacks) {
  if (node === null || node === void 0 || typeof node !== "object") return;
  if (Array.isArray(node)) {
    for (const child of node) {
      walk(child, callbacks);
    }
    return;
  }
  const record = node;
  const type = record.type;
  if (typeof type === "string" && type in callbacks) {
    callbacks[type](node);
  }
  for (const value of Object.values(record)) {
    if (value !== null && typeof value === "object") {
      walk(value, callbacks);
    }
  }
}

// ../../modules/file-parser/scope-analyzer/types.ts
var ScopeFlags = {
  Top: 1 << 0,
  Function: 1 << 1,
  Block: 1 << 2,
  Catch: 1 << 3,
  Class: 1 << 4,
  For: 1 << 5,
  Switch: 1 << 6,
  Module: 1 << 7,
  StrictMode: 1 << 8,
  ClassStaticBlock: 1 << 9,
  TsModuleBlock: 1 << 10,
  TsConditional: 1 << 11,
  Type: 1 << 12,
  Enum: 1 << 13
};

// ../../modules/file-parser/scope-analyzer/scopeAnalyzer.ts
import { parseSync } from "oxc-parser";
import visitorKeysDefault from "oxc-parser/src-js/generated/visit/keys.js";

// ../../modules/annotation/annotate-file.ts
function formatPropValue(value) {
  if (value === void 0) return null;
  if (typeof value === "function") return value.toString();
  return JSON.stringify(value);
}
function formatProps(props) {
  return Object.entries(props).map(([key, value]) => {
    const formatted = formatPropValue(value);
    if (formatted === null) return null;
    return ` ${key}={${formatted}}`;
  }).filter((s) => s !== null).join("");
}
function getExistingAttributeNames(element) {
  const names = /* @__PURE__ */ new Set();
  for (const attr of element.openingElement.attributes ?? []) {
    if (attr.type !== "JSXAttribute") continue;
    names.add(getTagName(attr.name));
  }
  return names;
}
function removeExistingProps(props, existingNames) {
  return Object.fromEntries(
    Object.entries(props).filter(([key]) => !existingNames.has(key))
  );
}
var FRAGMENT_TAG_NAMES = /* @__PURE__ */ new Set(["Fragment", "React.Fragment"]);
function annotateFile(source, filename, computeAnnotatedProps) {
  const parseResult = parseSync2(filename, source);
  const insertions = [];
  walk(parseResult.program, {
    JSXElement(node) {
      const element = node;
      const tagName = getTagName(element.openingElement.name);
      if (FRAGMENT_TAG_NAMES.has(tagName)) return;
      const props = removeExistingProps(
        computeAnnotatedProps({ element, tagName }),
        getExistingAttributeNames(element)
      );
      if (Object.keys(props).length === 0) return;
      const openingElement = element.openingElement;
      let insertPos = openingElement.end - 1;
      if (openingElement.selfClosing) {
        insertPos--;
        while (insertPos > openingElement.start && /\s/.test(source[insertPos - 1])) {
          insertPos--;
        }
      }
      insertions.push({ position: insertPos, text: formatProps(props) });
    }
  });
  insertions.sort((a, b) => a.position - b.position);
  const chunks = [];
  let lastPos = 0;
  for (const { position, text } of insertions) {
    chunks.push(source.slice(lastPos, position));
    chunks.push(text);
    lastPos = position;
  }
  chunks.push(source.slice(lastPos));
  return { annotatedSource: chunks.join(""), elementsAnnotated: insertions.length, parseResult };
}

// ../../modules/annotation/source-version.ts
function sourceVersionHash(content) {
  let djb2 = 5381;
  let sdbm = 0;
  for (let i = 0; i < content.length; i++) {
    const c = content.charCodeAt(i);
    djb2 = (djb2 << 5) + djb2 ^ c;
    sdbm = c + (sdbm << 6) + (sdbm << 16) - sdbm | 0;
  }
  return toHex8(djb2) + toHex8(sdbm);
}
function toHex8(n) {
  return (n >>> 0).toString(16).padStart(8, "0");
}

// index.ts
import { createFilter } from "@rollup/pluginutils";

// annotation-strategy.ts
import { realpathSync, statSync } from "fs";
import { relative } from "path";
var HTML_VOID_ELEMENTS = /* @__PURE__ */ new Set([
  "area",
  "base",
  "br",
  "col",
  "embed",
  "hr",
  "img",
  "input",
  "link",
  "meta",
  "param",
  "source",
  "track",
  "wbr"
]);
function isIntrinsicElement(tagName) {
  return tagName.length > 0 && tagName[0] === tagName[0].toLowerCase();
}
function hasJsxAttribute(info, attrName) {
  for (const attr of info.element.openingElement.attributes) {
    const name = attr.name;
    if (name?.name === attrName) return true;
  }
  return false;
}
function defaultComputeAnnotatedProps(info, filepath, modifiedtsMs, sourceVersion) {
  return {
    "data-tempo-filepath": filepath,
    "data-tempo-position": info.element.openingElement.start,
    "data-tempo-modifiedts": modifiedtsMs,
    // Content-hash identity of the file this element was annotated from. The
    // canvas reconcile loop equality-compares it against the write pipeline's
    // computed version to decide whether a capture reflects a given edit.
    "data-tempo-source-version": sourceVersion
  };
}
function computeTempoAnnotatedProps(params) {
  const {
    info,
    filepath,
    modifiedtsMs,
    sourceVersion,
    computeAnnotatedProps,
    elementSupportOverrides
  } = params;
  const userProps = computeAnnotatedProps ? computeAnnotatedProps(info, filepath, modifiedtsMs, sourceVersion) : defaultComputeAnnotatedProps(info, filepath, modifiedtsMs, sourceVersion);
  if (isIntrinsicElement(info.tagName)) {
    return {
      ...userProps,
      "data-tempo-supports-children": hasJsxAttribute(
        info,
        "data-tempo-supports-children"
      ) ? void 0 : !HTML_VOID_ELEMENTS.has(info.tagName)
    };
  }
  const override = elementSupportOverrides?.get(info.tagName);
  if (override) {
    return {
      ...userProps,
      "data-tempo-supports-children": hasJsxAttribute(
        info,
        "data-tempo-supports-children"
      ) ? void 0 : override.supportsChildren
    };
  }
  return userProps;
}

// index.ts
var DEFAULT_INCLUDE = ["**/*.tsx", "**/*.jsx"];
var DEFAULT_EXCLUDE = ["**/node_modules/**"];
var ROOT_META_NAME = "tempo-project-root";
function cleanModuleId(id) {
  return id.replace(/[?#].*$/, "");
}
function resolveCanonicalRoot(root) {
  try {
    return realpathSync2.native(root);
  } catch {
    return root;
  }
}
function hasRootMetaTag(html) {
  return /<meta[^>]+name=["']tempo-project-root["'][^>]*>/i.test(html);
}
function createErrorServer() {
  const errorsByFile = /* @__PURE__ */ new Map();
  function getMostRecentError() {
    if (errorsByFile.size === 0) return null;
    let latest = null;
    for (const err of errorsByFile.values()) latest = err;
    return latest;
  }
  function broadcastCurrentState() {
    const data = JSON.stringify(server.currentError);
    for (const client of server.clients) {
      client.write(`data: ${data}

`);
    }
  }
  const server = {
    get currentError() {
      return getMostRecentError();
    },
    clients: /* @__PURE__ */ new Set(),
    setError(error) {
      const key = error.file ?? error.loc?.file ?? "unknown";
      errorsByFile.delete(key);
      errorsByFile.set(key, error);
      broadcastCurrentState();
    },
    clearErrorForModules(updatedPaths) {
      if (errorsByFile.size === 0) return;
      let changed = false;
      for (const errorKey of [...errorsByFile.keys()]) {
        if (errorKey === "unknown" || updatedPaths.some((p) => errorKey.endsWith(p) || p.endsWith(errorKey))) {
          errorsByFile.delete(errorKey);
          changed = true;
        }
      }
      if (changed) broadcastCurrentState();
    },
    clearAll() {
      errorsByFile.clear();
      broadcastCurrentState();
    }
  };
  return server;
}
function readBody(req) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    req.on("data", (chunk) => chunks.push(chunk));
    req.on("end", () => resolve(Buffer.concat(chunks).toString()));
    req.on("error", reject);
  });
}
function handleErrorMiddleware(errorServer, req, res, next) {
  const url = req.url;
  if (!url?.startsWith("/__tempo/")) return next();
  if (url === "/__tempo/report-error" && req.method === "POST") {
    readBody(req).then((body) => {
      try {
        errorServer.setError(JSON.parse(body));
      } catch {
      }
      res.statusCode = 200;
      res.end("ok");
    });
    return;
  }
  if (url === "/__tempo/report-clear" && req.method === "POST") {
    errorServer.clearAll();
    res.statusCode = 200;
    res.end("ok");
    return;
  }
  if (url === "/__tempo/errors" && req.method === "OPTIONS") {
    res.writeHead(204, {
      "Access-Control-Allow-Origin": "*",
      "Access-Control-Allow-Methods": "GET, OPTIONS",
      "Access-Control-Allow-Headers": "*"
    });
    res.end();
    return;
  }
  if (url === "/__tempo/errors" && req.method === "GET") {
    res.writeHead(200, {
      "Content-Type": "text/event-stream",
      "Cache-Control": "no-cache",
      Connection: "keep-alive",
      "Access-Control-Allow-Origin": "*"
    });
    res.write(`data: ${JSON.stringify(errorServer.currentError)}

`);
    errorServer.clients.add(res);
    req.on("close", () => {
      errorServer.clients.delete(res);
    });
    return;
  }
  next();
}
function tempoVitePlugin(options = {}) {
  const enabled = !!process.env.TEMPO;
  const include = options.include === void 0 ? DEFAULT_INCLUDE : options.include;
  const exclude = options.exclude === void 0 ? DEFAULT_EXCLUDE : options.exclude;
  const filter = createFilter(include, exclude);
  let canonicalRoot = process.cwd();
  let command = "serve";
  const errorServer = createErrorServer();
  return {
    name: "tempo-annotate",
    enforce: "pre",
    // Expose error server for testing
    _tempoErrorServer: errorServer,
    configResolved(resolvedConfig) {
      if (!enabled) return;
      command = resolvedConfig.command;
      canonicalRoot = resolveCanonicalRoot(resolvedConfig.root);
    },
    configureServer(server) {
      if (!enabled) return;
      server.middlewares.use((req, res, next) => {
        handleErrorMiddleware(errorServer, req, res, next);
      });
      const interceptHmrSend = (target) => {
        const orig = target.send.bind(target);
        target.send = ((...args) => {
          const payload = args[0];
          if (typeof payload === "object" && payload !== null && "type" in payload) {
            const typed = payload;
            if (typed.type === "error") {
              const err = typed.err;
              errorServer.setError({
                type: "compile",
                message: err?.message ?? "Unknown error",
                stack: err?.stack,
                file: err?.id ?? err?.loc?.file,
                frame: err?.frame,
                plugin: err?.plugin,
                loc: err?.loc
              });
            } else if (typed.type === "update") {
              const updates = typed.updates;
              const paths = updates?.map((u) => u.path).filter((p) => typeof p === "string") ?? [];
              errorServer.clearErrorForModules(paths);
            }
          }
          return orig(...args);
        });
      };
      interceptHmrSend(server.ws);
      const clientHot = server.environments?.client?.hot;
      if (clientHot && clientHot !== server.ws) {
        interceptHmrSend(clientHot);
      }
    },
    transformIndexHtml(html) {
      if (!enabled || command !== "serve") {
        return;
      }
      if (hasRootMetaTag(html)) {
        return;
      }
      return [
        {
          tag: "meta",
          attrs: {
            name: ROOT_META_NAME,
            content: canonicalRoot
          },
          injectTo: "head"
        }
      ];
    },
    transform(source, id) {
      if (!enabled || command === "build") {
        return null;
      }
      if (id.startsWith("\0")) {
        return null;
      }
      const cleanId = cleanModuleId(id);
      if (!filter(cleanId)) {
        return null;
      }
      try {
        const filepath = relative2(canonicalRoot, cleanId);
        const modifiedtsMs = Math.trunc(statSync2(cleanId).mtimeMs);
        const sourceVersion = sourceVersionHash(source);
        const result = annotateFile(
          source,
          cleanId,
          (info) => computeTempoAnnotatedProps({
            info,
            filepath,
            modifiedtsMs,
            sourceVersion,
            computeAnnotatedProps: options.computeAnnotatedProps
          })
        );
        if (result.elementsAnnotated === 0) {
          return null;
        }
        return {
          code: result.annotatedSource,
          map: null
        };
      } catch {
        return null;
      }
    }
  };
}
var tempoAnnotate = tempoVitePlugin;
export {
  tempoAnnotate,
  tempoVitePlugin
};
